// @editor/dock/lib/utils/drag.js
'use strict';

const ipc = require('@base/electron-base-ipc');

const element = require('./element');

const pkg = require('../../package.json');
const ipcFlag = `${pkg.name}@${pkg.version}`;

let $mask = document.createElement('div');
$mask.style.position = 'absolute';
$mask.style.pointerEvents = 'none';
$mask.style.background = '#075692';
$mask.style.border = '2px solid #71c2ff';
$mask.style.opacity = 0.5;
$mask.style.zIndex = 1000;

let $line = document.createElement('div');
$line.style.position = 'absolute';
$line.style.pointerEvents = 'none';
$line.style.width = '2px';
$line.style.height = '28px';
$line.style.background = 'red';

let isDrag = false;

ipc.on(`${ipcFlag}:panel-drag-start`, () => {
    isDrag = true;
});

ipc.on(`${ipcFlag}:panel-drag-end`, () => {
    isDrag = false;
});

/**
 * 开始拖拽
 * @param event 
 * @param elem 
 */
let start = function (event, elem) {
    event.dataTransfer.setData('type', 'editor-panel');
    event.dataTransfer.setData('name', elem.name);

    // 记录被拖拽的布局元素的基本信息
    if (elem.$tabs.$groups) {
        event.dataTransfer.setData('panel-width', elem.$tabs.$groups.clientWidth);
        event.dataTransfer.setData('panel-height', elem.$tabs.$groups.clientHeight);
    }

    ipc.send(`${ipcFlag}:panel-drag-start`).callback(async (error, isDrop) => {
        if (error) {
            console.warn(error);
            ipc.send(`${ipcFlag}:panel-drag-closed`, false);
            return;
        }

        // 没有放到布局内，则不进行其他操作
        if (!isDrop) {
            ipc.send(`${ipcFlag}:panel-drag-closed`, false);
            return;
        }

        // 如果放到某个布局内，则尝试关闭现有的 panel
        let allow = await elem.$tabs.$groups.removePanel(elem.name, isDrop);
        if (allow === false) {
            ipc.send(`${ipcFlag}:panel-drag-closed`, false);
            return false;
        }
        ipc.send(`${ipcFlag}:panel-drag-closed`, true);
    });
};

/**
 * 拖拽结束
 * @param event 
 * @param elem 
 */
let end = function (event, elem) {
    setTimeout(() => {
        ipc.send(`${ipcFlag}:panel-drag-end`, false).callback((event, isClosed) => {
            if (!isClosed) {
                return;
            }
            element.refresh(elem);
        });
    }, 500);
};

/**
 * 放置面板到某个布局元素上
 * @param event 
 * @param elem 
 */
let drop = function (event, elem) {
    let box = elem.getBoundingClientRect();

    let x = event.pageX - box.left;
    let y = event.pageY - box.top;

    let sx = x / box.width;
    let sy = y / box.height;

    let type = '';

    if (sx < 0.3333) {
        if (sy > 0.3333 && sy < 0.6666) {
            type = 'left';
        }
    } else if (sx < 0.6666) {
        if (sy < 0.3333) {
            type = 'top';
        } else if (sy > 0.6666) {
            type = 'bottom'
        }
    } else {
        if (sy > 0.3333 && sy < 0.6666) {
            type = 'right';
        }
    }

    if (!type) {
        return false;
    }
    // event.preventDefault();
    // event.stopPropagation();

    const name = event.dataTransfer.getData('name');
    const height = event.dataTransfer.getData('panel-height');
    const width = event.dataTransfer.getData('panel-width');
    ipc.send(`${ipcFlag}:panel-drag-end`, true).callback((error, isClosed) => {
        if (!isClosed) {
            return;
        }

        // 添加新的布局
        element.addLayout(elem, {
            panels: [name],
        }, type, {
            width,
            height,
        });
        element.refresh(elem);
    });

    switchMask(false);
    return true;
};

/**
 * 
 * @param event 
 * @param elem 
 */
let dropTab = function (event, elem) {
    switchLine(false);
    let name = event.dataTransfer.getData('name');
    if (elem.name === name) {
        ipc.send(`${ipcFlag}:panel-drag-end`, false);
        return;
    }
    let $group = elem.$tabs.$groups;
    let index = $group.panels.indexOf(elem.name);

    let box = elem.getBoundingClientRect();
    let x = event.pageX - box.left;

    $line.style.top = `${box.top}px`;
    if (x / box.width >= 0.5) {
        index += 1;
    }

    ipc.send(`${ipcFlag}:panel-drag-end`, true).callback((error, isClosed) => {
        if (!isClosed) {
            return;
        }

        // 添加新的布局
        element.addPanel(elem, name, index);
        element.refresh(elem);
    });
};

/**
 * 判断显示线段
 * @param event 
 * @param elem 
 */
let tab = function (event, elem) {
    event.preventDefault();
    event.stopPropagation();

    let box = elem.getBoundingClientRect();
    let x = event.pageX - box.left;

    $line.style.top = `${box.top}px`;
    if (x / box.width < 0.5) {
        $line.style.left = `${box.left}px`;
    } else {
        $line.style.left = `${box.left + box.width}px`;
    }
    switchLine(true);
}

/**
 * 判断拖拽区域
 * @param event 
 * @param elem 
 */
let area = function (event, elem) {
    let box = elem.getBoundingClientRect();

    let x = event.pageX - box.left;
    let y = event.pageY - box.top;

    let sx = x / box.width;
    let sy = y / box.height;

    let type = '';

    if (sx < 0.3333) {
        if (sy > 0.3333 && sy < 0.6666) {
            type = 'left';

            $mask.style.left = box.left + 'px';
            $mask.style.top = box.top + 'px';
            $mask.style.width = box.width / 2 + 'px';
            $mask.style.height = box.height + 'px';
        }
    } else if (sx < 0.6666) {
        if (sy < 0.3333) {
            type = 'top';

            $mask.style.left = box.left + 'px';
            $mask.style.top = box.top + 'px';
            $mask.style.width = box.width + 'px';
            $mask.style.height = box.height / 2 + 'px';
        } else if (sy > 0.6666) {
            type = 'bottom'

            $mask.style.left = box.left + 'px';
            $mask.style.top = box.top + box.height / 2 + 'px';
            $mask.style.width = box.width + 'px';
            $mask.style.height = box.height / 2 + 'px';
        }
    } else {
        if (sy > 0.3333 && sy < 0.6666) {
            type = 'right';

            $mask.style.left = box.left + box.width / 2 + 'px';
            $mask.style.top = box.top + 'px';
            $mask.style.width = box.width / 2 + 'px';
            $mask.style.height = box.height + 'px';
        }
    }

    switchMask(!!type);

    return !!type;
};

/**
 * 切换 mask 的隐藏显示状态
 */
let switchMask = function (bool) {
    if (bool) {
        !$mask.parentElement && document.body.appendChild($mask);
    } else {
        $mask.parentElement && document.body.removeChild($mask);
    }
};

/**
 * 切换 line 的隐藏显示状态
 */
let switchLine = function (bool) {
    if (bool) {
        !$line.parentElement && document.body.appendChild($line);
    } else {
        $line.parentElement && document.body.removeChild($line);
    }
};

/**
 * 查询是否正在拖拽
 */
let getDrag = function () {
    return isDrag;
}

exports.start = start;
exports.end = end;
exports.drop = drop;
exports.dropTab = dropTab;
exports.area = area;
exports.tab = tab;
exports.switchMask = switchMask;
exports.switchLine = switchLine;
exports.getDrag = getDrag;
