// @editor/dock/lib/utils/layout.js
'use stirct';

const element = require('./element');
const Resize = require('../renderer/element/dock-resize');

let handler = {

    minimize(val) {
        return !!val;
    },

    percent (val) {
        val = val - 0;
        if (isNaN(val)) {
            val = 1;
        }
        return val;
    },

    minHeight (val) {
        val = val - 0;
        if (isNaN(val)) {
            val = 150;
        }
        return val;
    },

    minWidth (val) {
        val = val - 0;
        if (isNaN(val)) {
            val = 200;
        }
        return val;
    },

    direction (val) {
        if (
            val === 'column' ||
            val === 'row'
        ) {
            return val;
        }
        return 'none';
    },

    children (val) {
        if (!val || !Array.isArray(val) || val.length === 0) {
            return null;
        }
        return val;
    },
};

/**
 * 补全 layout json 数据
 */
let completion = function (layout) {
    // 补全默认值
    layout.minimize = handler.minimize(layout.minimize);
    layout.percent = handler.percent(layout.percent);
    layout.direction = handler.direction(layout.direction);
    layout['min-height'] = handler.minHeight(layout.minHeight);
    layout['min-width'] = handler.minHeight(layout.minWidth);
    layout.children = handler.children(layout.children);
    
    if (layout.children && layout.children.length > 0) {
        let percent = 0;
        let minWidth = 0;
        let minHeight = 0;

        if (layout.direction === 'column') {
            minWidth = layout['min-width'];
        } else if (layout.direction === 'row') {
            minHeight = layout['min-height'];
        }

        // 补全子 layout 数据
        layout.children.forEach((child) => {
            completion(child);
            // 如果布局最小化，则跳过记录百分比
            if (!child.minimize) {
                percent += child.percent;
            }
        });
        
        // 根据子节点的数据更新必要的数据
        layout.children.forEach((child) => {
            if (!child.minimize) {
                // children 的 percent 总值按比例改为 1
                child.percent /= percent;
            }

            // 根据子节点的宽高计算自己的最小宽高
            if (layout.direction === 'column') {
                minWidth = Math.max(minWidth, child['min-width']);
                minHeight += child['min-height'] + Resize.line;
            } else if (layout.direction === 'row') {
                minHeight = Math.max(minHeight, child['min-height']);
                minWidth += child['min-width'] + Resize.line;
            }
        });

        // 加上 resize 元素的宽高
        if (layout.direction === 'column') {
            minHeight -= Resize.line;
        } else if (layout.direction === 'row') {
            minWidth -= Resize.line;
        }

        layout['min-width'] = minWidth;
        layout['min-height'] = minHeight;
    }
};

class Snapshot {

    constructor ($resize, point) {
        this.reference = point;

        let $parent = $resize.parentElement;
        this.$parent = $parent;
        this.offset = 0;

        this.direction = $parent.direction;
        this.total = this.direction === 'column' ? $parent.offsetHeight : $parent.offsetWidth;
        // 排除点击元素自己的宽高
        this.total -= Resize.line;

        this.nextSiblingList = [];
        this.prevSiblingList = [];

        let $tmp = null;

        $tmp = $resize;
        while ($tmp.nextElementSibling) {
            $tmp = $tmp.nextElementSibling;
            if ($tmp.tagName !== 'DOCK-LAYOUT') {
                this.total -= Resize.line;
                continue;
            }
            if ($tmp.data.minimize) {
                this.total -= (this.direction === 'column' ? $tmp.offsetHeight : $tmp.offsetWidth);
                continue;
            }
            let flex = parseFloat($tmp.style.flex);
            this.nextSiblingList.push({
                element: $tmp,
                percent: flex,
                current: this.direction === 'column' ? $tmp.offsetHeight : $tmp.offsetWidth, // this.total * flex,
                min: parseInt(this.direction === 'column' ? $tmp.style.minHeight : $tmp.style.minWidth),
            });
        }

        $tmp = $resize;
        while ($tmp.previousElementSibling) {
            $tmp = $tmp.previousElementSibling;
            if ($tmp.tagName !== 'DOCK-LAYOUT') {
                this.total -= Resize.line;
                continue;
            }
            if ($tmp.data.minimize) {
                this.total -= (this.direction === 'column' ? $tmp.offsetHeight : $tmp.offsetWidth);
                continue;
            }
            let flex = parseFloat($tmp.style.flex);
            this.prevSiblingList.push({
                element: $tmp,
                percent: flex,
                current: this.direction === 'column' ? $tmp.offsetHeight : $tmp.offsetWidth, // this.total * flex,
                min: parseInt(this.direction === 'column' ? $tmp.style.minHeight : $tmp.style.minWidth),
            });
        }
    }
    
    translateNumber () {
        if (this.prevSiblingList.length === 0 || this.nextSiblingList.length === 0) {
            return;
        }

        let _total = this.total;
        this.prevSiblingList.forEach((item) => {
            let $elem = item.element;
            let num = Math.ceil(item.current);
            _total -= num;
            $elem.style.flex = `0 0  ${num}px`;
        });
        this.nextSiblingList.forEach((item) => {
            let $elem = item.element;
            let num = Math.ceil(item.current);
            _total -= num;
            if (_total < 0) {
                num += _total;
            }
            $elem.style.flex = `0 0  ${num}px`;
        });
    }

    translatePercent () {
        if (this.prevSiblingList.length === 0 || this.nextSiblingList.length === 0) {
            return;
        }

        this.nextSiblingList.forEach((item) => {
            let $elem = item.element;
            $elem.style.flex = `${item.percent}`;
        });
        this.prevSiblingList.forEach((item) => {
            let $elem = item.element;
            $elem.style.flex = `${item.percent}`;
        });
    }

    move (point) {
        if (this.prevSiblingList.length === 0 || this.nextSiblingList.length === 0) {
            return;
        }

        let offset = this.direction === 'column' ? point.y - this.reference.y : point.x - this.reference.x;
        this.offset = offset;

        let _offset;

        let array;
        if (offset > 0) {
            _offset = offset;
            array = [this.nextSiblingList, this.prevSiblingList];
        } else if (offset < 0) {
            _offset = -offset;
            array = [this.prevSiblingList, this.nextSiblingList];
        } else {
            return;
        }

        array.forEach((list, index) => {
            for (let i=0; i<list.length; i++) {
                let item = list[i];
    
                let num = item.current;

                if (index === 0) {
                    num -= _offset;
                } else {
                    num += _offset;
                }

                if (num < item.min) {
                    _offset = item.min - num;
                    num = item.min;

                    item.element.lock = true;
                    requestAnimationFrame(() => {
                        item.element.lock = false;
                        let event = document.createEvent('HTMLEvents');
                        event.initEvent('move', false, false);
                        item.element.dispatchEvent(event);
                    });
                } else {
                    _offset = 0;
                }

                if (item._num !== num && !item.element.lock) {
                    item.element.lock = true;
                    requestAnimationFrame(() => {
                        item.element.lock = false;
                        let event = document.createEvent('HTMLEvents');
                        event.initEvent('resize', false, false);
                        item.element.dispatchEvent(event);
                    });
                }

                item.element.style.flex = `0 0  ${num}px`;
                item._num = num;
            }
            _offset = Math.abs(offset) - _offset;
        });
    }

    apply () {
        if (this.prevSiblingList.length === 0 || this.nextSiblingList.length === 0) {
            return;
        }

        let total = this.total;
        this.nextSiblingList.forEach((item) => {
            item.current = item._num;
            item.percent = item.current / total;
            item.element.data.percent = item.percent;
        });
        this.prevSiblingList.forEach((item) => {
            item.current = item._num;
            item.percent = item.current / total;
            item.element.data.percent = item.percent;
        });

        element.emit(this.$parent);
    }
};

let _snapshotData = null;

/**
 * 快照当前的 resize 元素可以移动的所有元素信息
 * @param {*} $resize resize 元素
 */
let snapshot = function ($resize, x, y) {
    _snapshotData = new Snapshot($resize, {
        x, y,
    });
    _snapshotData.translateNumber();
};

/**
 * 拖拽之前的快照结果
 * @param {*} point 
 */
let dragLayout = function (point) {
    _snapshotData.move(point);
};

let clear = function () {
    _snapshotData.apply();
    _snapshotData.translatePercent();
};

exports.completion = completion;
exports.snapshot = snapshot;
exports.dragLayout = dragLayout;
exports.clear = clear;
