// @editor/setting/lib/browser/index.js
'use strict';

const ipc = require('@base/electron-base-ipc');
const ps = require('path'); // path system
const fs = require('fs');
const args = require('./args');

const { app } = require('electron');

const pkg = require('../../package.json');
const ipcFlag = `${pkg.name}@${pkg.version}`;

let name = app.getName();
let version = app.getVersion();
let argObject = args.analyze(process.argv);
const appPkg = require(ps.join(app.getAppPath(), 'package.json'));
if (argObject.dev === undefined) {
    argObject.dev = false;
} else {
    argObject.dev = !!argObject.dev;
}

// 如果没有传入 home 则定义一个默认的 home 地址
if (argObject.home && !ps.isAbsolute(argObject.home)) {
    argObject.home = ps.join(process.cwd(), argObject.project);
}
if (!argObject.home || !fs.existsSync(argObject.home)) {
    argObject.home = ps.join(app.getPath('home'), `.${name}` + (argObject.dev ? '_Develop' : ''));
}

// 如果有 project，并且是相对路径，则需要拼接实际路径
if (argObject.project && !ps.isAbsolute(argObject.project)) {
    argObject.project = ps.join(process.cwd(), argObject.project);
}

let PATH = {
    HOME: argObject.home,
    PROJECT: argObject.project || '',
    APP: process.cwd(),
};

module.exports = {
    name, version,
    title: appPkg.title,
    PATH,
    dev: !!argObject.dev,
    args: argObject,
};

ipc.on(`${ipcFlag}:query`, () => {
    return module.exports;
});
