// @cocos/creator-programming-rollup-plugin-mod-lo/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const url_1 = require("url");
const babel = __importStar(require("@babel/core"));
// NOTE: DO NOT REMOVE the null character `\0` as it may be used by other plugins
// e.g. https://github.com/rollup/rollup-plugin-node-resolve/blob/313a3e32f432f9eb18cc4c231cc7aac6df317a51/src/index.js#L74
const HELPERS = 'cce:/internal/rollupPluginModLoBabelHelpers.js';
function $({ modLo, origin, }) {
    const myName = 'rollup-plugin-mod-lo';
    const loadMetaMap = {};
    const originURL = origin !== null && origin !== void 0 ? origin : (0, url_1.pathToFileURL)(path_1.default.join(process.cwd(), path_1.default.sep));
    return {
        name: myName,
        resolveId,
        load,
    };
    async function resolveId(source, importer, _options) {
        if (source === HELPERS) {
            return {
                id: HELPERS,
            };
        }
        let resolved;
        try {
            if (!importer) {
                resolved = await modLo.resolve(source);
            }
            else {
                const loadMeta = loadMetaMap[importer];
                if (!loadMeta) {
                    return null;
                }
                const { url: importerURL, type: importerModLoType, } = loadMeta;
                resolved = await modLo.resolve(source, importerURL, importerModLoType);
            }
        }
        catch (err) {
            return this.error(err);
        }
        if (resolved.isExternal) {
            return {
                external: true,
                id: source,
            };
        }
        else {
            return {
                id: resolved.url.href,
            };
        }
    }
    async function load(id) {
        if (id === HELPERS) {
            // @ts-ignore
            return babel.buildExternalHelpers(null, 'module');
        }
        let url;
        try {
            url = new url_1.URL(id);
        }
        catch (_a) {
            return null;
        }
        let modLoModule;
        try {
            modLoModule = await modLo.load(url);
        }
        catch (err) {
            return null;
        }
        loadMetaMap[id] = {
            url,
            type: modLoModule.type,
        };
        const source = await modLoModule.module();
        return {
            code: source.code,
            map: source.map,
        };
    }
}
(function ($) {
    function filterWarns(warning, defaultHandler) {
        if (warning.code === 'CIRCULAR_DEPENDENCY') {
            if (warning.cycle && warning.cycle.every((id) => id.includes('node_modules'))) {
                return;
            }
        }
        defaultHandler(warning);
    }
    $.filterWarns = filterWarns;
    $.helperModule = HELPERS;
})($ || ($ = {}));
exports.default = $;
