// @editor/lib-programming/dist/mod-assembling/rollup-plugin-mod-lo/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generator_1 = __importDefault(require("@babel/generator"));
const path_1 = __importDefault(require("path"));
const url_1 = require("url");
const asserts_1 = require("../../utils/asserts");
const specifier_1 = require("../utils/specifier");
const babel = __importStar(require("@babel/core"));
// NOTE: DO NOT REMOVE the null character `\0` as it may be used by other plugins
// e.g. https://github.com/rollup/rollup-plugin-node-resolve/blob/313a3e32f432f9eb18cc4c231cc7aac6df317a51/src/index.js#L74
const HELPERS = '\0rollupPluginModLoBabelHelpers.js';
function $({ modLo, }) {
    const myName = 'rollup-plugin-mod-lo';
    const baseURL = url_1.pathToFileURL(process.cwd() + path_1.default.sep);
    const loadMetaMap = {};
    return {
        name: myName,
        resolveId,
        load,
    };
    async function resolveId(source, importer, _options) {
        if (source === HELPERS) {
            return {
                id: HELPERS,
            };
        }
        if (!importer) {
            asserts_1.asserts(!specifier_1.isRelativeSpecifier(source));
            return {
                id: source,
            };
        }
        const loadMeta = loadMetaMap[importer];
        if (!loadMeta) {
            return null;
        }
        const { url: importerURL, modLoMetadata: importerModLoMetadata, } = loadMeta;
        try {
            const resolved = await modLo.resolve(source, importerURL, importerModLoMetadata);
            if (resolved.isExternal) {
                return {
                    external: true,
                    id: source,
                };
            }
            else {
                return {
                    id: resolved.url.href,
                };
            }
        }
        catch (err) {
            return this.error(err);
        }
    }
    async function load(id) {
        if (id === HELPERS) {
            // @ts-ignore
            return babel.buildExternalHelpers(null, 'module');
        }
        let url;
        try {
            url = new url_1.URL(id);
        }
        catch (_a) {
            return null;
        }
        let modLoModule;
        try {
            modLoModule = await modLo.load(url);
        }
        catch (err) {
            return null;
        }
        loadMetaMap[id] = {
            url,
            modLoMetadata: modLoModule.modLoMetadata,
        };
        let code;
        let map;
        if (typeof modLoModule.code === 'string') {
            code = modLoModule.code;
            map = modLoModule.map;
        }
        else {
            const generateResult = generator_1.default(modLoModule.code);
            code = generateResult.code;
            map = generateResult.map;
        }
        return {
            code,
            map,
        };
    }
}
(function ($) {
    function filterWarns(warning, defaultHandler) {
        if (warning.code === 'CIRCULAR_DEPENDENCY') {
            if (warning.cycle && warning.cycle.every((id) => id.includes('node_modules'))) {
                return;
            }
        }
        defaultHandler(warning);
    }
    $.filterWarns = filterWarns;
    $.helperModule = HELPERS;
})($ || ($ = {}));
exports.default = $;
