// @editor/lib-programming/dist/mod-assembling/rollup-plugin-mod-lo/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generator_1 = __importDefault(require("@babel/generator"));
const path_1 = __importDefault(require("path"));
const url_1 = require("url");
const asserts_1 = require("../../utils/asserts");
const specifier_1 = require("../utils/specifier");
function $({ modLo, }) {
    const myName = 'rollup-plugin-mod-lo';
    const baseURL = url_1.pathToFileURL(process.cwd() + path_1.default.sep);
    const loadMetaMap = {};
    return {
        name: myName,
        resolveId,
        load,
    };
    async function resolveId(source, importer, _options) {
        if (!importer) {
            asserts_1.asserts(!specifier_1.isRelativeSpecifier(source));
            return {
                id: source,
            };
        }
        const loadMeta = loadMetaMap[importer];
        if (!loadMeta) {
            return null;
        }
        const { url: importerURL, modLoMetadata: importerModLoMetadata, } = loadMeta;
        try {
            const resolved = await modLo.resolve(source, importerURL, importerModLoMetadata);
            if (resolved.isExternal) {
                return {
                    external: true,
                    id: source,
                };
            }
            else {
                return {
                    id: resolved.url.href,
                };
            }
        }
        catch (err) {
            return this.error(err);
        }
    }
    async function load(id) {
        let url;
        try {
            url = new url_1.URL(id);
        }
        catch (_a) {
            return null;
        }
        let modLoModule;
        try {
            modLoModule = await modLo.load(url);
        }
        catch (err) {
            return null;
        }
        loadMetaMap[id] = {
            url,
            modLoMetadata: modLoModule.modLoMetadata,
        };
        let code;
        let map;
        if (typeof modLoModule.code === 'string') {
            code = modLoModule.code;
            map = modLoModule.map;
        }
        else {
            const generateResult = generator_1.default(modLoModule.code);
            code = generateResult.code;
            map = generateResult.map;
        }
        return {
            code,
            map,
        };
    }
}
(function ($) {
    function filterWarns(warning, defaultHandler) {
        if (warning.code === 'CIRCULAR_DEPENDENCY') {
            if (warning.cycle && warning.cycle.every((id) => id.includes('node_modules'))) {
                return;
            }
        }
        defaultHandler(warning);
    }
    $.filterWarns = filterWarns;
})($ || ($ = {}));
exports.default = $;
