// @cocos/creator-programming-quick-pack/lib/utils/chunk-io/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChunkIOBase = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
class ChunkIOBase {
    constructor({ chunkHomePath, }) {
        this._build = {
            chunks: {},
            entries: {},
        };
        this._chunkHomePath = path_1.default.normalize(chunkHomePath);
    }
    serializeRecord() {
        return this._build;
    }
    deserializeRecord(json) {
        this._build = json;
    }
    clear() {
        try {
            fs_extra_1.default.removeSync(this._chunkHomePath);
        }
        catch (err) {
        }
    }
    _calculateChunkCodeFileName(chunkId) {
        return path_1.default.join(this._chunkHomePath, this.calculateChunkCodeFileRelativePath(chunkId));
    }
    calculateChunkCodeFileRelativePath(chunkId) {
        return `${chunkId.slice(0, 2)}/${chunkId}.js`;
    }
    queryAllTimestamps() {
        const result = {};
        for (const [chunkId, { timestamp }] of Object.entries(this._build.chunks)) {
            const relativePath = this.calculateChunkCodeFileRelativePath(chunkId);
            result[relativePath] = timestamp;
        }
        return result;
    }
}
exports.ChunkIOBase = ChunkIOBase;
