// @base/electron-logger/lib/renderer.js
'use stirct';

const ipc = require('@base/electron-base-ipc');
const utils = require('./utils');
const { EventEmitter } = require('events');

const pkg = require('../package.json');
const ipcFlag = `${pkg.name}@${pkg.version}`;

class Logger extends EventEmitter {

    constructor () {
        super();
        this._logHandler = null;

        /**
         * 0 忽略所有记录
         * 1 只记录 error
         * 2 只记录 error 和 warn
         * 3 记录 error、warn、info
         * 4 记录 error、warn、info、log
         */
        this._level = 4;
    }

    /**
     * 设置一个 log 的自定义处理函数
     * 按照进程划分，所以这个数据不需要在所有进程间同步
     * @param {*} logPrefix 
     * @param {*} stackPrefix 
     */
    setLogHandler(func) {
        this._logHandler = func;
    }

    /**
     * 设置 logger 的记录等级
     * @param {number} level console 记录等级
     */
    setLevel (level) {
        if (typeof level !== 'number' || isNaN(level)) {
            level = 4;
        }
        this._level = level;
    }

    /**
     * 记录一个 log 信息
     * @param {*} type 
     * @param {*} message 
     */
    record (type, message) {
        const log = utils.formatMessage(type, message);
        if (this._logHandler) {
            this._logHandler(log);
        }
        ipc.send(`${ipcFlag}:record`, log);
    }

    /**
     * 查询 log
     * @param {*} option 
     * 
     * Logger.query({ regular: '*', type: 'error' });
     */
    query (option) {
        return new Promise((resolve, reject) => {
            ipc.send(`${ipcFlag}:query`, option).callback((error, list) => {
                if (error) {
                    return reject(error);
                }
                resolve(list);
            });
        });
    }

    /**
     * 清空所有的记录
     * 
     * @param {RegExp} regExp 匹配正则
     */
    clear (regExp) {
        ipc.send(`${ipcFlag}:clear`, regExp);
    }
}

let logger = module.exports = new Logger();

//////////////////////////////////////
// 截获常用的 console 方法，截获输出的信息

const vendor = {
    // level: develop
    log: console.log.bind(this),

    // level: release
    info: console.info.bind(this),
    warn: console.warn.bind(this),
    error: console.error.bind(this),
};

const electronReg = /^\(node:\d+\) electron:/;
console.log = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 3) {
        logger.record('log', utils.format(message, ...args));
    }
    vendor.log(message, ...args);
};

console.info = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 2) {
        logger.record('info', utils.format(message, ...args));
    }
    vendor.info(message, ...args);
};

console.warn = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 1) {
        logger.record('warn', utils.format(message, ...args));
    }
    vendor.warn(message, ...args);
};

console.error = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 0) {
        logger.record('error', utils.format(message, ...args));
    }
    vendor.error(message, ...args);
};

ipc.on(`${ipcFlag}:record`, (event, log) => {
    try {
        logger.emit('record', log);
    } catch (error) {
        vendor.error(error);
    }
});

ipc.on(`${ipcFlag}:clear`, (event) => {
    try {
        logger.emit('clear');
    } catch (error) {
        vendor.error(error);
    }
});
