// @editor/creator/static/theme/dark.js
'use strict';

// 配置文件版本号
exports.version = 1;

// 文字
exports.font = {
    normal: `BlinkMacSystemFont,'Helvetica Neue',Helvetica,'Lucida Grande','Segoe UI',Ubuntu,Cantarell,SourceHanSansCN-Normal,Arial,sans-serif`,
    editor: `Menlo,Monaco,"Courier New",monospace`,
};

// 颜色对比度比例
exports.contrast = 0.4;

// 渐变色参考值
exports.gradient = '#254850';

exports.color = {
    // //////////////////
    // 正常页面配色

    // 通常使用的配色
    normal: {
        fill: '#252830',     // 填充颜色
        contrast: '#a9adb9', // 对比色
        border: '#333846',   // 边框色
    },

    // 物体聚焦的时候的配色
    focus: {
        fill: '#196b8a',
        contrast: '#d9dde8',
        border: '#196b8a',
    },

    // 鼠标移入物体的配色
    hover: {
        fill: '#454955',
        contrast: '#d7d7d8',
        border: '#434857',
    },

    // 物体激活状态的时候的配色
    active: {
        fill: '#353945',
        contrast: '#d9dde8',
        border: '#434857',
    },

    // //////////////////
    // 按钮以及信息框配色

    // 重要的元素
    primary: {
        fill: '#1997c6',
        contrast: '#d9dde8',
        border: '#1997c6',
    },

    // 默认元素
    default: {
        fill: '#5a5f63',
        contrast: '#d6d5d5',
        border: '#6c757d',
    },

    // 显示成功消息的元素
    success: {
        fill: '#5cbf8c',
        contrast: '#eee',
        border: '#355a4e',
    },

    // 普通信息元素
    info: {
        fill: '#9f86ff',
        contrast: '#eee',
        border: '#9f86ff',
    },

    // 危险元素
    danger: {
        fill: '#e64759',
        contrast: '#eeeeee',
        border: '#653a3f',
    },

    // 警告元素
    warn: {
        fill: '#cab732',
        contrast: '#eee',
        border: '#e4d836',
    },
};

exports.size = {
    // 普通
    normal: {
        line: 24,    // 行高
        font: 12,    // 文字大小
        radius: 2,  // 边角半径
        border: 1,  // 边框
    },

    // 偏大
    big: {
        line: 26,
        font: 13,
        radius: 2,
        border: 1,
    },

    // 超大
    large: {
        line: 28,
        font: 14,
        radius: 3,
        border: 1,
    },

    // 小
    small: {
        line: 22,
        font: 11,
        radius: 1,
        border: 1,
    },

    // 很小
    little: {
        line: 20,
        font: 10,
        radius: 1,
        border: 1,
    },
};
