// @editor/creator/static/theme/light.js
'use strict';

exports.version = 1;

exports.font = {
    normal: `BlinkMacSystemFont,'Helvetica Neue',Helvetica,'Lucida Grande','Segoe UI',Ubuntu,Cantarell,SourceHanSansCN-Normal,Arial,sans-serif`,
    editor: `Menlo,Monaco,"Courier New",monospace`,
};

exports.contrast = -0.4;

exports.color = {
    // //////////////////
    // 正常页面配色

    // 普通
    normal: {
        fill: '#e0e0e0',
        contrast: '#333333',
        border: '#aaaaaa',
    },

    focus: {
        fill: '#77aaad',
        contrast: '#ffffff',
        border: '#13a6a0',
    },

    hover: {
        fill: '#cccccc',
        contrast: '#3b404c',
        border: '#16a4c2',
    },

    active: {
        fill: '#629da0',
        contrast: '#dddddd',
        border: '#13a6a0',
    },

    // //////////////////
    // 按钮以及信息框配色

    primary: {
        fill: '#1dc7ea',
        contrast: '#ffffff',
        border: '#1dc7ea',
    },

    default: {
        fill: '#ffffff',
        contrast: '#666666',
        border: '#bef5cb',
    },

    success: {
        fill: '#a5ffb9',
        contrast: '#525252',
        border: '#a2bf71',
    },

    info: {
        fill: '#ff9500',
        contrast: '#ffffff',
        border: '#ff9500',
    },

    danger: {
        fill: '#fb404b',
        contrast: '#ffffff',
        border: '#a97679',
    },

    warn: {
        fill: '#bb8911',
        contrast: '#ffffff',
        border: '#ffa534',
    },
};

exports.size = {
    // 普通
    normal: {
        line: 24,    // 行高
        font: 12,    // 文字大小
        radius: 4,  // 边角半径
        border: 1,  // 边框
    },

    // 偏大
    big: {
        line: 28,
        font: 14,
        radius: 3,
        border: 1,
    },

    // 超大
    large: {
        line: 32,
        font: 16,
        radius: 4,
        border: 1,
    },

    // 小
    small: {
        line: 20,
        font: 10,
        radius: 2,
        border: 1,
    },

    // 很小
    little: {
        line: 16,
        font: 8,
        radius: 1,
        border: 1,
    },
};
