// @editor/quick-compiler/test/run.js
// @ts-check

const { QuickCompiler } = require('../dist/index');
const ps = require('path');
const logUpdate = require('log-update');
const { Stage } = require('../dist/progress');

module.exports.run = async (inputDir, outputDir) => {
    const baseOutDir = outputDir;
    const previewOutDir = ps.join(baseOutDir, 'preview');
    const editorOutDir = ps.join(baseOutDir, 'editor');

    const targets = [
        {
            dir: editorOutDir,
            format: 'commonjs',
            // inlineSourceMap: true,
            // indexedSourceMap: true,
            usedInElectron509: true,
            targets: 'Electron 5.0.8',
            featureUnitPrefix: 'cce:/internal/x/cc-fu/',
            includeIndex: {
                features: [
                    'base',
                    'gfx-webgl',
                    'gfx-webgl2',
                    '3d',
                    '2d',
                    'ui',
                    'particle',
                    'particle-2d',
                    'physics-framework',
                    'physics-2d-framework',
                    'intersection-2d',
                    'primitive',
                    'profiler',
                    'audio',
                    'video',
                    'terrain',
                    'webview',
                    'tween',
                    'tiled-map',
                    'spine',
                    'dragon-bones',
                ],
            },
        },
        {
            dir: previewOutDir,
            format: 'systemjs',
            loose: true,
            featureUnitPrefix: 'cce:/internal/x/cc-fu/',
        },
    ];

    const log = logUpdate.create(process.stdout, {
        showCursor: true,
    });

    const targetProgresses = targets.map(() => '');

    console.time('build');
    await (new QuickCompiler({
        rootDir: inputDir,
        outDir: baseOutDir,
        targets,
        logFile: 'log.txt',
        onProgress: (target, message) => {
            if (message.stage === Stage.transform) {
                const percentage = (message.progress / message.total * 100).toFixed(2);
                targetProgresses[target] =
                    `[QuickCompiler] ${target} Transforming... ${message.progress}/${message.total}(${percentage}%) ${message.file}`;
            } else if (message.stage === Stage.bundle) {
                targetProgresses[target] =
                    `[QuickCompiler] ${target} Bundling...`;
            }
            log(targetProgresses.join('\n'));
        },
    }).build());
    console.timeEnd('build');
};
