// @editor/build-systemjs/dist/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rollup = __importStar(require("rollup"));
const rollup_plugin_terser_1 = require("rollup-plugin-terser");
const path_1 = __importDefault(require("path"));
// @ts-ignore
const plugin_multi_entry_1 = __importDefault(require("@rollup/plugin-multi-entry"));
const plugin_replace_1 = __importDefault(require("@rollup/plugin-replace"));
const staticDir = path_1.default.join(__dirname, '..', 'static');
const systemJsSourceDir = path_1.default.join(staticDir, 'systemjs-source');
async function build(options) {
    const { domScriptLoad, amd, } = options;
    const plugins = [
        plugin_multi_entry_1.default(),
        plugin_replace_1.default({
            TRACING: 'false',
        }),
    ];
    if (options.terser) {
        plugins.push(rollup_plugin_terser_1.terser());
    }
    const inputs = [
        path_1.default.join(systemJsSourceDir, 'src', 'system-core.js'),
    ];
    if (domScriptLoad) {
        inputs.push(path_1.default.join(systemJsSourceDir, 'src', 'features', 'script-load.js'));
    }
    else {
        inputs.push(path_1.default.join(staticDir, 'cc', 'forbid-script-load.js'));
    }
    inputs.push(path_1.default.join(staticDir, 'cc', 'patches.js'), path_1.default.join(systemJsSourceDir, 'src', 'extras', 'named-register.js'));
    if (amd) {
        inputs.push(path_1.default.join(systemJsSourceDir, 'src', 'extras', 'amd.js'));
    }
    const rollupBuild = await rollup.rollup({
        input: inputs,
        plugins: plugins,
    });
    await rollupBuild.write({
        format: 'iife',
        strict: false,
        file: options.outFile,
        sourcemap: options.sourceMap,
    });
}
exports.build = build;
