// @editor/asset-db/libs/dependency.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssociatedFiles = exports.DependencyManager = void 0;
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const migrator_1 = require("./migrator");
const migrations = [{
        version: '1.0.0',
        migrate: async (cacheInfo, manager) => {
            const recordInfo = {
                data: {
                    path: {},
                    uuid: {},
                },
                version: DependencyManager.version,
            };
            Object.keys(cacheInfo.path).forEach((path) => {
                recordInfo.data.path[(0, path_1.relative)(manager.pathRoot, path)] = cacheInfo.path[path].map((subPath => (0, path_1.relative)(manager.pathRoot, subPath)));
            });
            Object.keys(cacheInfo.uuid).forEach((path) => {
                recordInfo.data.uuid[(0, path_1.relative)(manager.pathRoot, path)] = cacheInfo.uuid[path];
            });
            return recordInfo;
        }
    }];
function getDefaultRecordInfo() {
    return {
        data: {
            path: {},
            uuid: {},
        },
        version: DependencyManager.version,
    };
}
// 关联性 map，一个资源更改后需要通知依赖这个资源的其他资源
// 这里记录的就是依赖某个资源的其他资源列表
const associatedMap = {};
/**
 * 资源关联以及依赖关系列表，主要影响导入队列以及是否需要重新导入
 * 部分数据需要固化到硬盘上
 */
class DependencyManager {
    constructor(customConsole, pathRoot) {
        // 依赖列表，一个对象依赖的所有对象的名字数组
        // { uuid: { 'id': [] }, url: { 'db://test/1.json': [] }, path: { '/Users/xxx': [] } }
        this.dependMap = getDefaultRecordInfo().data;
        // 保存使用的 timer
        this._saveTimer = null;
        this.console = customConsole || console;
        this.pathRoot = pathRoot;
    }
    /**
     * 设置用于记录的 json 文件
     * @param json
     */
    async setRecordJSON(path) {
        this.file = path;
        try {
            await this._restoreCache(path);
        }
        catch (error) {
            this.console.warn(error);
        }
    }
    async _restoreCache(path) {
        const cacheInfo = await this.readRecordJSON(path);
        if (!cacheInfo) {
            return;
        }
        // 处理缓存数据
        const { path: pathRecord, uuid: uuidRecord } = this.dependMap;
        Object.keys(cacheInfo.data.path).forEach((relativePath) => {
            pathRecord[(0, path_1.join)(this.pathRoot, relativePath)] = cacheInfo.data.path[relativePath].map((subPath => (0, path_1.join)(this.pathRoot, subPath)));
        });
        Object.keys(cacheInfo.data.uuid).forEach((relativePath) => {
            uuidRecord[(0, path_1.join)(this.pathRoot, relativePath)] = cacheInfo.data.uuid[relativePath];
        });
        // 补全关联列表
        for (let type in this.dependMap) {
            const typeMap = this.dependMap[type];
            for (let path in typeMap) {
                const array = typeMap[path];
                array.forEach((urlOrPathOrUUID) => {
                    associatedMap[urlOrPathOrUUID] = associatedMap[urlOrPathOrUUID] || [];
                    if (associatedMap[urlOrPathOrUUID].indexOf(path) !== -1) {
                        return;
                    }
                    associatedMap[urlOrPathOrUUID].push(path);
                });
            }
        }
    }
    async readRecordJSON(path) {
        if (!(0, fs_extra_1.existsSync)(path)) {
            return;
        }
        try {
            const cacheInfo = (0, fs_extra_1.readJSONSync)(path);
            const version = cacheInfo.version ? cacheInfo.version : '0.0.0';
            const migrator = new migrator_1.Migrator(migrations, version, {
                onError: (error, stage, data, ...args) => {
                    this.console.warn(`Migrate error in dependencyManager`);
                    this.console.warn(error);
                }
            });
            return await migrator.run(cacheInfo, DependencyManager.version, [this]);
        }
        catch (error) {
            this.console.error(error);
            return;
        }
    }
    /*
     * 延迟保存依赖文件
     */
    save() {
        clearTimeout(this._saveTimer);
        this._saveTimer = setTimeout(() => {
            this.saveImmediate();
        }, 400);
    }
    /*
     * 立即保存依赖文件
     */
    saveImmediate() {
        clearTimeout(this._saveTimer);
        if (this.file) {
            const recordInfo = getDefaultRecordInfo();
            // 存储到文件系统里需要记录相对路径
            const { path: pathRecord, uuid: uuidRecord } = this.dependMap;
            Object.keys(pathRecord).forEach((path) => {
                recordInfo.data.path[(0, path_1.relative)(this.pathRoot, path)] = pathRecord[path].map((file) => (0, path_1.relative)(this.pathRoot, file));
            });
            Object.keys(uuidRecord).forEach((path) => {
                recordInfo.data.uuid[(0, path_1.relative)(this.pathRoot, path)] = uuidRecord[path];
            });
            (0, fs_extra_1.outputJSONSync)(this.file, recordInfo, { spaces: 2 });
        }
    }
    /**
     * 记录一个资源依赖的所有资源列表
     * 允许传入 url、uuid、path 三种依赖的格式
     * @param path
     * @param dependNames
     */
    add(type, key, depends) {
        const typeMap = this.dependMap[type] = this.dependMap[type] || {};
        const array = typeMap[key] = typeMap[key] || [];
        if (!Array.isArray(depends)) {
            depends = [depends];
        }
        depends.forEach((depend) => {
            if (depend && array.indexOf(depend) !== -1) {
                return;
            }
            // 记录到依赖列表
            array.push(depend);
            // 记录到关联列表
            const associated = associatedMap[depend] = associatedMap[depend] || [];
            if (associated.indexOf(key) === -1) {
                associated.push(key);
            }
        });
        this.save();
    }
    /**
     * 清空一个资源的依赖记录
     * @param name
     */
    remove(type, key) {
        if (!this.dependMap[type] || !this.dependMap[type][key]) {
            return;
        }
        // 删除依赖列表
        this.dependMap[type][key].forEach((str) => {
            const array = associatedMap[str];
            const index = array.indexOf(key);
            if (index !== -1) {
                array.splice(index, 1);
            }
            if (array.length === 0) {
                delete associatedMap[str];
            }
        });
        delete this.dependMap[type][key];
        this.save();
    }
    /**
     * 销毁一个依赖管理器实例
     * @param manager
     */
    destroy() {
        // 将管理器记录的所有数据从全局缓存内删除
        // 但不清除缓存 json，因为之后还可能需要使用
        for (let type in this.dependMap) {
            const typeMap = this.dependMap[type];
            for (let key in typeMap) {
                const array = typeMap[key];
                array.forEach((str) => {
                    if (!associatedMap[str]) {
                        return;
                    }
                    const index = associatedMap[str].indexOf(key);
                    if (index !== -1) {
                        associatedMap[str].splice(index, 1);
                    }
                    if (associatedMap[str].length === 0) {
                        delete associatedMap[str];
                    }
                });
            }
        }
    }
}
exports.DependencyManager = DependencyManager;
DependencyManager.version = '1.0.0';
/**
 * 获取被影响的资源列表
 * @param urlOrPathOrUUID
 */
function getAssociatedFiles(urlOrPathOrUUID) {
    return associatedMap[urlOrPathOrUUID] || [];
}
exports.getAssociatedFiles = getAssociatedFiles;
