// @editor/asset-db/libs/dependency.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssociatedFiles = exports.DependencyManager = void 0;
const fs_extra_1 = require("fs-extra");
// 关联性 map，一个资源更改后需要通知依赖这个资源的其他资源
// 这里记录的就是依赖某个资源的其他资源列表
const associatedMap = {};
/**
 * 资源关联以及依赖关系列表
 * 部分数据需要固化到硬盘上
 */
class DependencyManager {
    constructor(customConsole) {
        // 依赖列表，一个对象依赖的所有对象的名字数组
        // { uuid: { 'id': [] }, url: { 'db://test/1.json': [] }, path: { '/Users/xxx': [] } }
        this.dependMap = {};
        // 保存使用的 timer
        this._saveTimer = null;
        this.console = customConsole || console;
    }
    /**
     * 设置用于记录的 json 文件
     * @param json
     */
    async setRecordJSON(json) {
        this.file = json;
        if ((0, fs_extra_1.existsSync)(json)) {
            try {
                this.dependMap = (0, fs_extra_1.readJSONSync)(this.file);
            }
            catch (error) {
                this.console.error(error);
                this.dependMap = {};
            }
        }
        else {
            this.dependMap = {};
        }
        // 补全关联列表
        for (let type in this.dependMap) {
            const typeMap = this.dependMap[type];
            for (let path in typeMap) {
                const array = typeMap[path];
                array.forEach((urlOrPathOrUUID) => {
                    associatedMap[urlOrPathOrUUID] = associatedMap[urlOrPathOrUUID] || [];
                    if (associatedMap[urlOrPathOrUUID].indexOf(path) !== -1) {
                        return;
                    }
                    associatedMap[urlOrPathOrUUID].push(path);
                });
            }
        }
    }
    /*
     * 延迟保存依赖文件
     */
    save() {
        clearTimeout(this._saveTimer);
        this._saveTimer = setTimeout(() => {
            this.saveImmediate();
        }, 400);
    }
    /*
     * 立即保存依赖文件
     */
    saveImmediate() {
        clearTimeout(this._saveTimer);
        this.file && (0, fs_extra_1.outputJSONSync)(this.file, this.dependMap, { spaces: 2 });
    }
    /**
     * 记录一个资源依赖的所有资源列表
     * 允许传入 url、uuid、path 三种依赖的格式
     * @param path
     * @param dependNames
     */
    add(type, key, depends) {
        const typeMap = this.dependMap[type] = this.dependMap[type] || {};
        const array = typeMap[key] = typeMap[key] || [];
        if (!Array.isArray(depends)) {
            depends = [depends];
        }
        depends.forEach((depend) => {
            if (depend && array.indexOf(depend) !== -1) {
                return;
            }
            // 记录到依赖列表
            array.push(depend);
            // 记录到关联列表
            const associated = associatedMap[depend] = associatedMap[depend] || [];
            if (associated.indexOf(key) === -1) {
                associated.push(key);
            }
        });
        this.save();
    }
    /**
     * 清空一个资源的依赖记录
     * @param name
     */
    remove(type, key) {
        if (!this.dependMap[type] || !this.dependMap[type][key]) {
            return;
        }
        // 删除依赖列表
        this.dependMap[type][key].forEach((str) => {
            const array = associatedMap[str];
            const index = array.indexOf(key);
            if (index !== -1) {
                array.splice(index, 1);
            }
            if (array.length === 0) {
                delete associatedMap[str];
            }
        });
        delete this.dependMap[type][key];
        this.save();
    }
    /**
     * 销毁一个依赖管理器实例
     * @param manager
     */
    destroy() {
        // 将管理器记录的所有数据从全局缓存内删除
        // 但不清除缓存 json，因为之后还可能需要使用
        for (let type in this.dependMap) {
            const typeMap = this.dependMap[type];
            for (let key in typeMap) {
                const array = typeMap[key];
                array.forEach((str) => {
                    if (!associatedMap[str]) {
                        return;
                    }
                    const index = associatedMap[str].indexOf(key);
                    if (index !== -1) {
                        associatedMap[str].splice(index, 1);
                    }
                    if (associatedMap[str].length === 0) {
                        delete associatedMap[str];
                    }
                });
            }
        }
    }
}
exports.DependencyManager = DependencyManager;
/**
 * 获取被影响的资源列表
 * @param urlOrPathOrUUID
 */
function getAssociatedFiles(urlOrPathOrUUID) {
    return associatedMap[urlOrPathOrUUID] || [];
}
exports.getAssociatedFiles = getAssociatedFiles;
