// @editor/asset-db/libs/info.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoManager = void 0;
const fs_extra_1 = require("fs-extra");
/**
 * 缓存所有文件的 mtimeMs 时间，用于比对是否修改
 * 这部分数据需要落地到文件系统
 */
class InfoManager {
    constructor(customConsole) {
        this.recordInfo = {
            version: InfoManager.version,
            map: {},
            missing: {},
        };
        // 保存使用的 timer
        this._saveTimer = null;
        this.console = customConsole || console;
    }
    /**
     * 设置记录数据的 json 文件
     * @param json
     */
    setRecordJSON(json) {
        this.file = json.replace('.json', `${InfoManager.version}.json`);
        // 优先取当前版本号的配置数据
        if ((0, fs_extra_1.existsSync)(this.file)) {
            try {
                const recordInfo = (0, fs_extra_1.readJSONSync)(this.file);
                this.loadRecordJson(recordInfo);
            }
            catch (error) {
                this.console.error(error);
            }
            return;
        }
        // 兼容获取旧版本数据, 同时写入到新的数据文件内
        if ((0, fs_extra_1.existsSync)(json)) {
            try {
                const recordInfo = (0, fs_extra_1.readJSONSync)(json);
                this.loadRecordJson(recordInfo);
                (0, fs_extra_1.outputJSONSync)(this.file, this.recordInfo, { spaces: 2 });
            }
            catch (error) {
                this.console.error(error);
            }
        }
    }
    loadRecordJson(recordInfo) {
        if (!recordInfo.version) {
            // 旧版本数据需要做一次整理，将 missing 数据整理出来
            Object.keys(recordInfo).forEach((path) => {
                const info = recordInfo[path];
                if (info.missing) {
                    delete info.missing;
                    info.uuid && (this.recordInfo.missing[info.uuid] = {
                        path,
                        time: info.time,
                        removeTime: Date.now(),
                    });
                }
                else {
                    delete info.missing;
                    this.recordInfo.map[path] = info;
                }
            });
        }
        else {
            this.recordInfo = recordInfo;
        }
    }
    /**
     * 销毁一个管理器实例
     * @param manager
     */
    destroy() {
        this.recordInfo.map = {};
        this.recordInfo.missing = {};
    }
    /*
     * 延迟保存依赖文件
     */
    save() {
        this._saveTimer && clearTimeout(this._saveTimer);
        this._saveTimer = setTimeout(() => {
            this.saveImmediate();
        }, 400);
    }
    /*
     * 立即保存记录文件
     */
    saveImmediate() {
        this._saveTimer && clearTimeout(this._saveTimer);
        this.file && (0, fs_extra_1.outputJSONSync)(this.file, this.recordInfo, { spaces: 2 });
    }
    /**
     * 更新一个缓存数据
     * @param path
     * @param mtimeMs
     * @param uuid
     */
    add(path, mtimeMs, uuid) {
        if (this.recordInfo.map[path]
            && this.recordInfo.map[path].uuid === uuid
            && this.recordInfo.map[path].time === mtimeMs) {
            return;
        }
        if (uuid) {
            this.recordInfo.map[path] = {
                time: mtimeMs,
                uuid: uuid,
            };
        }
        else {
            this.recordInfo.map[path] = {
                time: mtimeMs,
            };
        }
        this.save();
    }
    /**
     * 删除缓存的一个 mtime 数据
     * @param path
     */
    remove(path) {
        if (!this.recordInfo.map[path]) {
            return;
        }
        const info = this.recordInfo.map[path];
        this.addMissing(path, info);
        delete this.recordInfo.map[path];
        this.save();
    }
    /**
     * 获取缓存的 stats 对象
     * @param path
     */
    get(path) {
        return this.recordInfo.map[path] || null;
    }
    /**
     * 添加一个丢失的资源信息
     * @param path
     * @param info
     */
    addMissing(path, info) {
        info.uuid && (this.recordInfo.missing[info.uuid] = {
            path,
            time: info.time,
            removeTime: Date.now(),
        });
    }
    /**
     * 根据 uuid 获取丢失的资源信息
     * @param uuid
     * @returns
     */
    getMissingInfo(uuid) {
        return this.recordInfo.missing[uuid] || null;
    }
    /**
     * 对比现在文件和内存里缓存的 stats 是否有修改
     * 返回是否相等
     * @param path
     * @param stats
     */
    compare(path, mtimeMs) {
        // 如果缓存不存在，则认为修改了
        const target = this.recordInfo.map[path];
        if (!target) {
            return false;
        }
        // 如果 ms 记录相等，则认为没有修改
        if (target.time === mtimeMs) {
            return true;
        }
        return false;
    }
    async forEach(handler) {
        for (const path in this.recordInfo.map) {
            await handler(path, this.recordInfo.map[path]);
        }
    }
    ;
}
exports.InfoManager = InfoManager;
InfoManager.version = '1.0.0';
