// @editor/asset-db/libs/info.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoManager = void 0;
const fs_extra_1 = require("fs-extra");
/**
 * 缓存所有文件的 mtimeMs 时间，用于比对是否修改
 * 这部分数据需要落地到文件系统
 */
class InfoManager {
    constructor() {
        // 列表
        this.map = {};
        // 保存使用的 timer
        this._saveTimer = null;
    }
    /**
     * 设置记录数据的 json 文件
     * @param json
     */
    async setRecordJSON(json) {
        this.file = json;
        if ((0, fs_extra_1.existsSync)(json)) {
            try {
                this.map = (0, fs_extra_1.readJSONSync)(this.file);
            }
            catch (error) {
                console.error(error);
                this.map = {};
            }
        }
        else {
            this.map = {};
        }
    }
    /**
     * 销毁一个管理器实例
     * @param manager
     */
    destroy() {
        this.map = {};
    }
    /*
     * 延迟保存依赖文件
     */
    save() {
        clearTimeout(this._saveTimer);
        this._saveTimer = setTimeout(() => {
            this.saveImmediate();
        }, 400);
    }
    /*
     * 立即保存依赖文件
     */
    saveImmediate() {
        clearTimeout(this._saveTimer);
        this.file && (0, fs_extra_1.outputJSONSync)(this.file, this.map, { spaces: 2 });
    }
    /**
     * 更新一个缓存数据
     * @param path
     * @param mtimeMs
     * @param uuid
     */
    add(path, mtimeMs, uuid) {
        if (this.map[path]
            && this.map[path].uuid === uuid
            && this.map[path].time === mtimeMs) {
            return;
        }
        if (uuid) {
            this.map[path] = {
                time: mtimeMs,
                uuid: uuid,
                missing: false,
            };
        }
        else {
            this.map[path] = {
                time: mtimeMs,
            };
        }
        this.save();
    }
    /**
     * 删除缓存的一个 mtime 数据
     * @param path
     */
    remove(path) {
        if (!this.map[path]) {
            return;
        }
        delete this.map[path];
        this.save();
    }
    /**
     * 获取缓存的 stats 对象
     * @param path
     */
    get(path) {
        return this.map[path] || null;
    }
    /**
     * 对比现在文件和内存里缓存的 stats 是否有修改
     * 返回是否相等
     * @param path
     * @param stats
     */
    compare(path, mtimeMs) {
        // 如果缓存不存在，则认为修改了
        const target = this.map[path];
        if (!target) {
            return false;
        }
        // 如果 ms 记录相等，则认为没有修改
        if (target.time === mtimeMs) {
            return true;
        }
        return false;
    }
}
exports.InfoManager = InfoManager;
