// @cocos/creator-programming-quick-pack/lib/middleware.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickPackMiddleware = void 0;
const path_1 = __importDefault(require("path"));
const url_1 = require("url");
const constants_1 = require("./constants");
const proper_lockfile_1 = __importDefault(require("proper-lockfile"));
class QuickPackMiddleware {
    constructor(workspace) {
        this._workspace = workspace;
        this._workspaceURL = (0, url_1.pathToFileURL)(path_1.default.join(workspace, path_1.default.sep));
    }
    get workspace() {
        return this._workspace;
    }
    get workspaceURL() {
        return this._workspaceURL;
    }
    get mainRecordPath() {
        return path_1.default.join(this._workspace, 'main-record.json');
    }
    get assemblyRecordPath() {
        return path_1.default.join(this._workspace, 'assembly-record.json');
    }
    get chunkHomePath() {
        return (0, url_1.fileURLToPath)(new url_1.URL(constants_1.CHUNK_HOME_RELATIVE_URL, this._workspaceURL));
    }
    get importMapPath() {
        return (0, url_1.fileURLToPath)(new url_1.URL(constants_1.IMPORT_MAP_RELATIVE_URL, this._workspaceURL));
    }
    get resolutionDetailMapPath() {
        return (0, url_1.fileURLToPath)(new url_1.URL(constants_1.RESOLUTION_DETAIL_MAP_RELATIVE_URL, this._workspaceURL));
    }
    async lock(maxRetryTime = Infinity) {
        return proper_lockfile_1.default.lock(this._workspace, { retries: { maxRetryTime }, stale: 5000 });
    }
    async unlock() {
        return proper_lockfile_1.default.unlock(this._workspace);
    }
}
exports.QuickPackMiddleware = QuickPackMiddleware;
