// @editor/lib-programming/dist/mod-assembling/quick-pack/loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickPackLoaderContext = exports.QuickPackLoader = exports.ResourceNotFoundError = void 0;
const loader_context_1 = require("./utils/loader-context");
Object.defineProperty(exports, "QuickPackLoaderContext", { enumerable: true, get: function () { return loader_context_1.LoaderContext; } });
const base_1 = require("./utils/chunk-io/base");
const asserts_1 = require("../../utils/asserts");
const chunk_url_1 = require("./utils/chunk-url");
const url_1 = require("url");
const path_1 = __importDefault(require("path"));
const specifier_1 = require("../utils/specifier");
const url_2 = require("../utils/url");
const i18n_1 = require("../../utils/i18n");
class ResourceNotFoundError extends Error {
    constructor(id) {
        super(i18n_1.i18nTranslate('quick_pack_loader_resource_not_found_error', { id }));
    }
}
exports.ResourceNotFoundError = ResourceNotFoundError;
class QuickPackLoader extends base_1.ChunkIOBase {
    constructor(context, options) {
        super({ chunkRecordFile: context.chunkRecordFile, sourceCacheDir: context.sourceCacheDir });
        this._importMap = {};
        this._entries = [];
        this._origin = context.origin;
        this._options = options;
        const realFsRoot = url_1.pathToFileURL(this._origin).pathname;
        this._fsLocator = new chunk_url_1.FsLocator(realFsRoot);
    }
    async getImportMap() {
        return this._importMap;
    }
    getEntries() {
        return this._entries;
    }
    async getResource(link) {
        const resourceId = this.getResourceId(link);
        return await this.getResourceFromId(resourceId);
    }
    getResourceId(link) {
        const normalizedLink = this._normalizeLink(link);
        const quickPackURL = linkToQuickPackURL(normalizedLink);
        const chunkId = this._fsLocator.toChunkId(quickPackURL);
        return chunkId;
    }
    async getResourceFromId(resourceId) {
        let chunkId;
        let isMap;
        if (resourceId.endsWith('.map')) {
            isMap = true;
            chunkId = resourceId.substr(0, resourceId.length - 4);
        }
        else {
            isMap = false;
            chunkId = resourceId;
        }
        const chunk = this._chunks[chunkId];
        if (!chunk) {
            throw new ResourceNotFoundError(resourceId);
        }
        const physicalLocation = chunk.physicalLocation;
        const file = this._getChunkSourceFile(physicalLocation);
        return {
            type: 'file',
            path: isMap ? `${file}.map` : file,
        };
    }
    /**
     * 获取指定资源的 mtime 时间戳。若不存在则返回负值。
     */
    async getResourceMTimestamp(resource) {
        var _a, _b;
        return (_b = (_a = this._chunks[this._getChunkId(resource)]) === null || _a === void 0 ? void 0 : _a.mTimestamp) !== null && _b !== void 0 ? _b : -1;
    }
    /**
     * 获取指定所有资源的 mtime 时间戳。不存在的资源将返回负值。
     */
    async getResourceMTimestamps(resources) {
        return resources.map((resourceId) => { var _a, _b; return (_b = (_a = this._chunks[this._getChunkId(resourceId)]) === null || _a === void 0 ? void 0 : _a.mTimestamp) !== null && _b !== void 0 ? _b : -1; });
    }
    async reload() {
        await this._loadChunkRecord();
        ({
            importMap: this._importMap,
            entries: this._entries,
        } = await this._createImportMap(this._origin, this._options.importMapBasedRoot));
    }
    async _createImportMap(origin, importMapBasedRoot) {
        var _a, _b, _c;
        asserts_1.asserts(importMapBasedRoot.endsWith('/'));
        const fsLocator = this._fsLocator;
        const entries = [];
        const imports = {};
        const scopes = {};
        const getImportMapBasedUrl = (url) => {
            if (chunk_url_1.isQuickPackURL(url)) {
                return `${importMapBasedRoot}${quickPackURLToExternalLink(url)}`;
            }
            else {
                // All other urls are not import map arrived.
                return url.href;
            }
        };
        const normalizeSpecifier = (specifier, baseUrl) => {
            var _a;
            // https://wicg.github.io/import-maps/#resolve-an-imports-match
            if (specifier_1.isRelativeSpecifier(specifier)) {
                return url_2.tryParseURL(specifier, baseUrl);
            }
            else {
                return (_a = url_2.tryParseURL(specifier)) !== null && _a !== void 0 ? _a : specifier;
            }
        };
        const originDirUrl = url_1.pathToFileURL(`${origin}${path_1.default.sep}`);
        const originNpURL = new url_1.URL(fsLocator.getOrigin().href);
        if (!originNpURL.pathname.endsWith('/')) {
            originNpURL.pathname += '/';
        }
        imports[originDirUrl.href] = getImportMapBasedUrl(originNpURL);
        const urlResolves = {};
        for (const [chunkId, chunk] of Object.entries(this._chunks)) {
            const chunkURL = fsLocator.toQuickPackURL(chunkId);
            const myUrlImportMapBased = getImportMapBasedUrl(chunkURL);
            if (chunk.isEntry) {
                entries.push(myUrlImportMapBased);
            }
            if (chunk.exposeAs) {
                imports[chunk.exposeAs] = myUrlImportMapBased;
            }
            const scopeCurMod = myUrlImportMapBased;
            for (const [specifier, resolvedChunkId] of Object.entries(chunk.imports)) {
                const resolvedUrl = fsLocator.toQuickPackURL(resolvedChunkId);
                const normalizedSpecifier = normalizeSpecifier(specifier, chunkURL);
                // If the specifier is erroneous, continue.
                // Otherwise, the normalized specifier is an URL or bare specifier.
                if (!normalizedSpecifier) {
                    continue;
                }
                const isNormal = typeof normalizedSpecifier !== 'string' &&
                    normalizedSpecifier.href === resolvedUrl.href;
                const resolveKey = typeof normalizedSpecifier === 'string'
                    ? normalizedSpecifier
                    : getImportMapBasedUrl(normalizedSpecifier);
                const resolves = (_a = urlResolves[resolveKey]) !== null && _a !== void 0 ? _a : (urlResolves[resolveKey] = { normal: false });
                if (isNormal) {
                    resolves.normal = true;
                }
                else {
                    const specialResolves = (_b = resolves.specialResolves) !== null && _b !== void 0 ? _b : (resolves.specialResolves = {});
                    specialResolves[scopeCurMod] = resolvedUrl;
                }
            }
        }
        const scopeThis = scopes[getImportMapBasedUrl(chunk_url_1.quickPackRootURL)] = {};
        for (const [specifier, { normal, specialResolves }] of Object.entries(urlResolves)) {
            if (!specialResolves) {
                // If there is no special match, do nothing
                continue;
            }
            if (normal) {
                scopeThis[specifier] = specifier;
            }
            const allResolves = Object.entries(specialResolves);
            if (!normal && allResolves.every(([, resolved], _index, [[, firstResolved]]) => resolved.href === firstResolved.href)) {
                const [, resolved] = allResolves[0];
                scopeThis[specifier] = getImportMapBasedUrl(resolved);
            }
            else {
                for (const [importer, resolved] of allResolves) {
                    ((_c = scopes[importer]) !== null && _c !== void 0 ? _c : (scopes[importer] = {}))[specifier] = getImportMapBasedUrl(resolved);
                }
            }
        }
        const sortObject = (o, fn = (v) => v) => Object.keys(o).sort().reduce((r, k) => (r[k] = fn(o[k]), r), {});
        const importMap = {
            imports: sortObject(imports),
            scopes: sortObject(scopes, (map) => sortObject(map)),
        };
        return {
            entries,
            importMap,
        };
    }
    _normalizeLink(link) {
        return link.startsWith('/') ? link : `/${link}`;
    }
    _getChunkId(resourceId) {
        return resourceId.endsWith('.map')
            ? resourceId.substr(0, resourceId.length - 4)
            : resourceId;
    }
}
exports.QuickPackLoader = QuickPackLoader;
function quickPackURLToExternalLink(url) {
    return `${url.pathname.substr(1)}${url.search}`;
}
function linkToQuickPackURL(link) {
    return new url_1.URL(link, chunk_url_1.quickPackRootURL);
}
