// @editor/lib-programming/dist/mod-assembling/quick-pack/loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickPackLoaderContext = exports.QuickPackLoader = exports.ResourceNotFoundError = void 0;
const loader_context_1 = require("./utils/loader-context");
Object.defineProperty(exports, "QuickPackLoaderContext", { enumerable: true, get: function () { return loader_context_1.LoaderContext; } });
const base_1 = require("./utils/chunk-io/base");
const chunk_url_1 = require("./utils/chunk-url");
const url_1 = require("url");
const i18n_1 = require("../../utils/i18n");
const fs_extra_1 = __importDefault(require("fs-extra"));
class ResourceNotFoundError extends Error {
    constructor(id) {
        super(i18n_1.i18nTranslate('quick_pack_loader_resource_not_found_error', { id }));
    }
}
exports.ResourceNotFoundError = ResourceNotFoundError;
class QuickPackLoader extends base_1.ChunkIOBase {
    constructor(context, options) {
        super({ chunkRecordFile: context.chunkRecordFile, sourceCacheDir: context.sourceCacheDir });
        this._importMap = {};
        this._physicalLocationMap = {};
        this._origin = context.origin;
        this._options = options;
        this._importMapFile = context.importMapFile;
        const realFsRoot = url_1.pathToFileURL(this._origin).pathname;
        this._fsLocator = new chunk_url_1.FsLocator(realFsRoot);
    }
    async getImportMap() {
        return this._importMap;
    }
    async getResource(link) {
        const resourceId = this.getResourceId(link);
        return await this.getResourceFromId(resourceId);
    }
    getResourceId(link) {
        return link.startsWith('/') ? link.substr(1) : link;
    }
    async getResourceFromId(resourceId) {
        const physicalLocation = resourceId;
        const file = this._getChunkSourceFile(physicalLocation);
        return {
            type: 'file',
            path: file,
        };
    }
    /**
     * 获取指定资源的 mtime 时间戳。若不存在则返回负值。
     */
    async getResourceMTimestamp(resource) {
        var _a, _b;
        const physicalLocation = resource;
        return (_b = (_a = this._physicalLocationMap[physicalLocation]) === null || _a === void 0 ? void 0 : _a.mTimestamp) !== null && _b !== void 0 ? _b : -1;
    }
    /**
     * 获取指定所有资源的 mtime 时间戳。不存在的资源将返回负值。
     */
    async getResourceMTimestamps(resources) {
        return resources.map((resourceId) => { var _a, _b; return (_b = (_a = this._physicalLocationMap[resourceId]) === null || _a === void 0 ? void 0 : _a.mTimestamp) !== null && _b !== void 0 ? _b : -1; });
    }
    async reload() {
        await this._loadChunkRecord();
        this._physicalLocationMap = {};
        for (const [chunkId, chunk] of Object.entries(this._chunks)) {
            this._physicalLocationMap[chunk.physicalLocation] = {
                mTimestamp: chunk.mTimestamp,
            };
        }
        this._importMap = await fs_extra_1.default.readJson(this._importMapFile);
    }
}
exports.QuickPackLoader = QuickPackLoader;
function quickPackURLToExternalLink(url) {
    return `${url.pathname.substr(1)}${url.search}`;
}
function linkToQuickPackURL(link) {
    return new url_1.URL(link, chunk_url_1.quickPackRootURL);
}
