// @base/electron-logger/lib/browser/index.js
'use stirct';

const ipc = require('@base/electron-base-ipc');
const utils = require('../utils');
const { EventEmitter } = require('events');

const pkg = require('../../package.json');
const ipcFlag = `${pkg.name}@${pkg.version}`;

class Logger extends EventEmitter {

    constructor() {
        super();

        this._logHandler = null;

        this._prefix = {
            log: '',
            stack: '',
        };

        /**
         * 0 忽略所有记录
         * 1 只记录 error
         * 2 只记录 error 和 warn
         * 3 记录 error、warn、info和log
         */
        this._level = 4;
        this._logs = [];
    }

    /**
     * 设置一个 log 的自定义处理函数
     * 按照进程划分，所以这个数据不需要在所有进程间同步
     * @param {*} logPrefix 
     * @param {*} stackPrefix 
     */
    setLogHandler(func) {
        this._logHandler = func;
    }

    /**
     * 设置 log 记录等级
     * 
     */
    setLevel(level) {
        if (typeof level !== 'number' || isNaN(level)) {
            level = 3;
        }
        this._level = level;
    }

    /**
     * 记录一个 log 信息
     * @param {*} type 
     * @param {*} message 
     */
    record(type, message) {
        let log = utils.formatMessage(type, message);

        if (this._logHandler) {
            if (this._logHandler(log) === false) {
                return;
            }
        }

        this._logs.push(log);

        try {
            this.emit('record', log);
        } catch (error) {
            vendor.error(error);
        }
        ipc.broadcast(`${ipcFlag}:record`, log);
    }

    /**
     * 查询 log
     * @param {*} option 
     * 
     * todo:
     * Logger.query({ regular: '*', type: 'error' });
     * Logger.query({ sort: 'desc', num: 10, });
     */
    query(option) {
        return this._logs;
    }

    /**
     * 清空所有的记录
     * 
     * @param {RegExp} regExp 匹配正则
     */
    clear(regExp) {
        if (regExp && regExp instanceof RegExp) {
            for (let i = 0; i < this._logs.length; i++) {
                if (regExp.test(this._logs[i].message)) {
                    this._logs.splice(i--, 1);
                }
            }
        } else {
            this._logs = [];
        }

        try {
            this.emit('clear');
        } catch (error) {
            vendor.error(error);
        }
        ipc.broadcast(`${ipcFlag}:clear`);
    }
}

let logger = module.exports = new Logger();

//////////////////////////////////////
// 截获常用的 console 方法，截获输出的信息

const vendor = {
    // level: develop
    log: console.log.bind(console),

    // level: release
    info: console.info.bind(console),
    warn: console.warn.bind(console),
    error: console.error.bind(console),
};

const electronReg = /^(\(node:\d+\) electron:)/;
console.log = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 2) {
        logger.record('log', utils.format(message, ...args));
    }
    vendor.log(message, ...args);
};

console.info = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 2) {
        logger.record('info', utils.format(message, ...args));
    }
    vendor.info(message, ...args);
};

console.warn = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 1) {
        logger.record('warn', utils.format(message, ...args));
    }
    vendor.warn(message, ...args);
};

console.error = function (message, ...args) {
    if (electronReg.test(message)) {
        vendor.warn(message, ...args);
        return;
    }
    if (logger._level > 0) {
        logger.record('error', utils.format(message, ...args));
    }
    vendor.error(message, ...args);
};

/////////////////////
// 渲染进程的 ipc 反馈

// 渲染进程请求记录 log 信息
ipc.on(`${ipcFlag}:record`, (event, log) => {
    logger._logs.push(log);
    logger.emit('record', log);
    ipc.broadcast(`${ipcFlag}:record`, log);
    try {
        if (log.stack) {
            // TODO 此处应由 vStacks.decode 实现
            const err = new Error(log.message);
            const logName = log.type[0].toLocaleUpperCase() + log.type.slice(1);
            err.stack = log.stack.replace(new RegExp('^' + err.name), logName);
            err.name = logName;
            vendor[log.type] && vendor[log.type](err);
            return;
        }
        vendor[log.type] && vendor[log.type](log.message);
    } catch (error) {
        vendor.warn(error);
    }
});

// 渲染进程请求数据
ipc.on(`${ipcFlag}:query`, (event, option) => {
    const list = logger.query(option);
    event.reply(null, list);
    return list;
});

// 渲染进清空所有的记录
ipc.on(`${ipcFlag}:clear`, (event) => {
    return logger.clear();
});
