// @editor/extension-sdk/sdk/manager/download/lib/CacheManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cacheManage = exports.CacheManager = void 0;
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const config_1 = require("../../config");
class CacheManager {
    storeName(info) {
        const { name, version } = info;
        return `${name}@${version}.zip`;
    }
    async get(info) {
        const root = (0, config_1.getCacheStoreFolder)();
        const storePath = (0, path_1.join)(root, this.storeName(info));
        if (await (0, fs_extra_1.pathExists)(storePath))
            return storePath;
        return undefined;
    }
    async save(zipPath, info) {
        const root = (0, config_1.getCacheStoreFolder)();
        await (0, fs_extra_1.ensureDir)(root);
        const storePath = (0, path_1.join)(root, this.storeName(info));
        if (!(await (0, fs_extra_1.pathExists)(storePath)))
            await (0, fs_extra_1.copy)(zipPath, storePath);
    }
    async remove(info) {
        const root = (0, config_1.getCacheStoreFolder)();
        const storePath = (0, path_1.join)(root, this.storeName(info));
        if (await (0, fs_extra_1.pathExists)(storePath))
            await (0, fs_extra_1.remove)(storePath);
    }
    async clean() {
        const root = (0, config_1.getCacheStoreFolder)();
        if (await (0, fs_extra_1.pathExists)(root))
            await (0, fs_extra_1.remove)(root);
    }
}
exports.CacheManager = CacheManager;
exports.cacheManage = new CacheManager();
