// @editor/dock/lib/renderer/element/dock-tab.js
"use strict";

const ps = require("path");
const i18n = require("@base/electron-i18n");
const panel = require("@editor/panel");

const utils = require("../../utils");

let $tabs = [];

/**
 * 更新一个 tab 与 panel 信息相关的数据
 * @param {*} $elem
 */
let updateInfo = function ($elem) {
  // 查询 panel 的 info
  let info = panel.queryInfo($elem.name);

  // $img 存在的情况
  if ($elem.$img) {
    if (info && info.userData && info.userData.icon) {
      // panel 定义了 icon
      let dirname = info.userData.packageDir || info.dirname;
      $elem.$img.innerHTML = `<img src="${ps.join(
        dirname,
        info.userData.icon
      )}" />`;
    } else {
      // panel 没定义 icon，显示默认的 svg
      $elem.$img.innerHTML = '';
      // `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512">
      //     <path d="M369.9 97.9L286 14C277 5 264.8-.1 252.1-.1H48C21.5 0 0 21.5 0 48v416c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48V131.9c0-12.7-5.1-25-14.1-34zM332.1 128H256V51.9l76.1 76.1zM48 464V48h160v104c0 13.3 10.7 24 24 24h104v288H48z"/>
      // </svg>`;
    }
  }

  if (info && info.userData && info.userData.title) {
    // 如果定义了 title，则显示 title
    let title = info.userData.title;
    if (title.startsWith("i18n:")) {
      title = title.substr(5);
      title = i18n.translation(title) || title;
    }
    $elem.$name.innerHTML = title;
  } else {
    // 没有定义则显示 panel 的 name
    $elem.$name.innerHTML = $elem.data.name;
  }
};

class DockTab extends window.HTMLElement {
  constructor() {
    super();

    this.data = {
      name: "", // panel 的名字
      title: "", // tab 显示的标题，这个可能和 panel 的 name 是不同的
      active: false,
    };

    this.$tabs = null; // 上一级的 DOCK-TABS 元素

    this.addEventListener("click", (event) => {
      event.stopPropagation();
      event.preventDefault();

      this.$tabs.$groups.active = this.name;
    });

    this.addEventListener("dblclick", () => {
      const $layout = this.$tabs.$groups.parentElement;
      if (
        !$layout ||
        !$layout.parentElement ||
        $layout.parentElement.tagName !== "DOCK-LAYOUT"
      ) {
        return;
      }
      const minimize = !$layout.minimize;
      const children = $layout.parentElement.children;
      // 判断是否还有父层级的 layout，如果有，则判断是否所有的 layout 都最小化了（最后一个不允许最小化）
      if (
        minimize &&
        Array.prototype.every.call(children, ($elem) => {
          if ($elem.tagName !== "DOCK-LAYOUT" || $layout === $elem) {
            return true;
          }
          return $elem.minimize;
        })
      ) {
        return;
      }
      $layout.minimize = minimize;

      if (!minimize) {
        Array.prototype.forEach.call(children, ($elem) => {
          if ($elem.tagName !== "DOCK-LAYOUT" || $layout === $elem) {
            return true;
          }
          return ($elem.data.percent *= 1 - $layout.data.percent);
        });
      }
    });

    this.addEventListener("dragover", (event) => {
      event.preventDefault();
      event.stopPropagation();
      utils.drag.tab(event, this);
    });

    this.addEventListener("dragleave", (event) => {
      event.preventDefault();
      event.stopPropagation();
      utils.drag.switchLine(false);
    });

    this.addEventListener("dragstart", (event) => {
      utils.drag.start(event, this);
    });

    this.addEventListener("dragend", (event) => {
      utils.drag.end(event, this);
      utils.drag.switchMask(false);
    });

    this.addEventListener("drop", (event) => {
      event.preventDefault();
      event.stopPropagation();
      utils.drag.dropTab(event, this);
    });
  }

  connectedCallback() {
    $tabs.push(this);

    this.setAttribute("draggable", "true");

    this.$img = document.createElement("span");
    this.$img.classList.add("img");
    this.insertBefore(this.$img, this.firstElementChild);

    updateInfo(this);
  }

  disconnectedCallback() {
    let index = $tabs.indexOf(this);
    $tabs.splice(index, 1);

    this.removeChild(this.$img);
  }

  get name() {
    return this.data.name;
  }

  set name(val) {
    this.data.name = val;

    if (!this.$name) {
      this.$name = document.createElement("span");
      this.appendChild(this.$name);
    }

    updateInfo(this);
  }

  get title() {
    return this._title;
  }

  set title(val) {
    this._title = val;
  }

  get active() {
    return this.data.active;
  }

  set active(val) {
    val = !!val;
    this.data.active = val;
    if (val) {
      this.setAttribute("active", val);
    } else {
      this.removeAttribute("active");
    }
  }
}

module.exports = DockTab;

// 初始化的时候可能 panel 还没有注册，需要监听事件自动更新数据
panel.on("register", (name) => {
  $tabs.forEach(($tab) => {
    if ($tab.name === name) {
      updateInfo($tab);
    }
  });
});

panel.on("change", (name) => {
  $tabs.forEach(($tab) => {
    if ($tab.name === name) {
      updateInfo($tab);
    }
  });
});

// i18n 切换的时候更新数据
i18n.on("switch", () => {
  $tabs.forEach(($tab) => {
    updateInfo($tab);
  });
});
