// @editor/dock/lib/renderer/element/dock-tabs.js
'use strict';

const panel = require('@editor/panel');
const menu = require('@base/electron-menu');

const utils = require('../../utils');

const tabsHTML = `
<nav></nav>
<div>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512">
        <path d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"></path>
    </svg>
</div>`;

let menuParser = null;

function createDockTab(dockTabs, name) {
    let $tab = document.createElement('dock-tab');
    $tab.name = name;
    $tab.active = $tab.name === dockTabs.active;
    $tab.$tabs = dockTabs;
    $tab.addEventListener('mouseup', async (event) => {
        if (event.button === 2) {
            dockTabs.$groups.active = $tab.name;
            await popupMenu(event, dockTabs);
        }
    });
    return $tab;
}

async function popupMenu(event, dockTabs) {
    const name = dockTabs.data.active;
    const $groups = event.currentTarget.parentElement.parentElement.parentElement;
    const templates = menuParser ? await menuParser($groups, name) : [];
    menu.popup({
        menu: templates,
    });
}

class DockTabs extends window.HTMLElement {

    static registerMenuParser(func) {
        menuParser = func;
    }

    constructor() {
        super();

        // groups custom element
        this.$groups = null;

        this.data = {
            panels: [],
            active: '',
        };

        this.$nav = null;
        this.$more = null;

        this.addEventListener('dragover', (event) => {
            utils.drag.tab(event, this.$nav.lastElementChild);
        });

        this.addEventListener('dragleave', (event) => {
            utils.drag.switchLine(false);
        });

        // this.addEventListener('dragend', (event) => {
            // utils.drag.end(event, this.$nav.lastElementChild);
            // utils.drag.switchMask(false);
        // });

        this.addEventListener('drop', (event) => {
            event.preventDefault();
            event.stopPropagation();
            utils.drag.dropTab(event, this.$nav.lastElementChild);
        });
    }

    connectedCallback() {
        if (!this.$nav) {
            this.innerHTML = tabsHTML;
            this.$nav = this.querySelector('nav');
            this.$more = this.querySelector('svg');

            for (let i = 0; i < this.data.panels.length; i++) {
                const $tab = createDockTab(this, this.data.panels[i]);
                this.$nav.appendChild($tab);
            }

            this.$more = this.querySelector('div > svg');
            this.$more.addEventListener('click', async (event) => {
                await popupMenu(event, this);
            });
        }
    }

    disconnectedCallback() {
        // this.innerHTML = '';
        // this.$nav = null;
        // this.$more = null;
    }

    get panels() {
        return this.data.panels;
    }

    set panels(val) {
        // this.data.panels = val;

        for (let i = 0; i < val.length; i++) {
            this.addTab(val[i]);
        }
    }

    get active() {
        return this.data.active;
    }

    set active(val) {
        this.data.active = val;

        if (!this.$nav) {
            return;
        }

        let $tabs = this.$nav.children;
        for (let i = 0; i < $tabs.length; i++) {
            let $tab = $tabs[i];
            $tab.active = $tab.name === val;
        }
    }

    /**
     * 添加一个 tab 按钮
     * @param {*} name tab 的名字
     * @param {*} index 索引
     */
    addTab(name, index) {
        let panels = this.data.panels;
        if (index === undefined || index >= panels.length) {
            index = panels.length;
        }

        if (panels.indexOf(name) !== -1) {
            return;
        }

        panels.splice(index, 0, name);

        if (!this.$nav) {
            return;
        }

        const $tab = createDockTab(this, name);
        let $children = this.$nav.children;
        if (index >= $children.length) {
            this.$nav.appendChild($tab);
        } else {
            this.$nav.insertBefore($tab, $children[index]);
        }
    }

    /**
     * 移除一个 tab 按钮
     * @param {*} name 名字
     */
    removeTab(name) {
        let panels = this.data.panels;
        let index = panels.indexOf(name);
        if (index === -1) {
            return;
        }
        panels.splice(index, 1);

        if (!this.$nav) {
            return;
        }

        let $tabs = this.$nav.children;
        for (let i = 0; i < $tabs.length; i++) {
            let $tab = $tabs[i];
            if ($tab.name === name) {
                this.$nav.removeChild($tab);
            }
        }
    }
}

module.exports = DockTabs;
