// @editor/dock/lib/renderer/element/dock-frame.js
'use strict';

const fs = require('fs');
const utils = require('../../utils');

const style = fs.readFileSync(`${__dirname}/index.css`, 'utf8');
let customStyle = '';

let $frames = [];

class DockFrame extends window.HTMLElement {

    /**
     * 传入一个 css 地址，加载的时候将引入这个 css
     * @param {string} path 
     */
    static importStyle (path) {
        if (!fs.existsSync(path)) {
            return;
        }
        customStyle = fs.readFileSync(path, 'utf8');

        $frames.forEach(($frame) => {
            let $style = $frame.shadowRoot.querySelector('#custom-style');
            if ($style) {
                let html = fs.readFileSync(path, 'utf-8');
                $style.innerHTML = html;
            }
        });
    }

    constructor () {
        super();
        this.attachShadow({
            mode: 'open'
        });

        this.shadowRoot.innerHTML = `
        <style>${style}</style>
        <style id="custom-style">${customStyle}</style>
        <slot></slot>
        `;

        this.$layout = null;
    }

    connectedCallback () {
        $frames.push(this);
    }

    disconnectedCallback () {
        let index = $frames.indexOf(this);
        $frames.splice(index, 1);
    }

    /**
     * 应用布局数据
     * @param {*} json 
     */
    layout (json) {
        // todo 版本兼容
        let layout = json.layout;
        
        if (!this.$layout) {
            this.$layout = document.createElement('dock-layout');
            this.shadowRoot.appendChild(this.$layout);
            this.$layout.$parent = this;
        }

        if (layout.percent !== 1) {
            layout.percent = 1;
        }

        utils.layout.completion(layout);

        this.$layout.layout(layout);
    }

    /**
     * 在一个已有的面板上，添加一个面板
     * @param {*} existsPanel 
     * @param {*} appendPanel 
     * @param {*} position 
     */
    appendPanel(existsPanel, appendPanel, position) {
        utils.element.appendPanel(this, existsPanel, appendPanel);
    }

    /**
     * 获取布局数据
     */
    dump () {
        let layout = this.$layout.dump();
        return {
            version: 1,
            layout: layout,
        };
    }
}

module.exports = DockFrame;
