// @editor/dock/lib/renderer/element/dock-groups.js
'use strict';

const panel = require('@editor/panel');
const utils = require('../../utils');

class DockGroups extends window.HTMLElement {

    constructor() {
        super();

        this.data = {
            panels: [],
            active: '',
            isConnect: false,
        };

        this.$tabs = null;
        this.$panels = null;

        // panels 的最小尺寸更新的时候触发，需要更新 groups 的最小尺寸
        this.addEventListener('min-size-resize', (event) => {
            if (event.minWidth) {
                this.minWidth = event.minWidth;
            }
            if (event.minHeight) {
                this.minHeight = event.minHeight;
            }
            utils.element.refresh(this);
        });

        this._resizeFn = function () {
            Array.prototype.forEach.call(this.$panels.children, (child) => {
                let event = document.createEvent('HTMLEvents');
                event.initEvent('resize', false, false);
                child.dispatchEvent(event);
            });
        }.bind(this);

        this.addEventListener('move', function () {
            Array.prototype.forEach.call(this.$panels.children, (child) => {
                let event = document.createEvent('HTMLEvents');
                event.initEvent('move', false, false);
                child.dispatchEvent(event);
            });
        });

        this.addEventListener('resize', this._resizeFn);

        this.addEventListener('focus', (event) => {
            this._focused = true;
            this._onFocus(event);
        });

        this.addEventListener('blur', (event) => {
            this._focused = false;
            this._onBlur(event);
        });

        this.addEventListener('panel-focus', (event) => {
            this.active = event.target.name;
            this._focused = true;
            this._onFocus(event);
        });
        this.addEventListener('panel-blur', (event) => {
            this._focused = false;
            this._onBlur(event);
        });
        this.addEventListener('click', () => {
            if (this.$panels) {
                this.$panels.active = this.active;
            }
        });
    }

    get panels() {
        return this.data.panels;
    }

    set panels(array) {
        this.data.panels = array;
    }

    get active() {
        return this.data.active;
    }

    set active(val) {
        this.data.active = val;
        if (this.$tabs) {
            this.$tabs.active = val;
        }
        if (this.$panels) {
            this.$panels.active = val;
        }

        if (this.data.isConnect) {
            utils.element.emit(this);
        }
    }

    connectedCallback() {
        if (!this.$tabs) {
            this.$tabs = document.createElement('dock-tabs');
            this.appendChild(this.$tabs);
            // this.$tabs.panels = Array.prototype.concat(this.data.panels);
            this.$tabs.$groups = this;
        }
        this.$tabs.active = this.data.active;

        this.setAttribute('tabindex', '-1');

        if (!this.$panels) {
            this.$panels = document.createElement('dock-panels');
            this.appendChild(this.$panels);
            this.$panels.active = this.data.active;
            this.$panels.panels = Array.prototype.concat(this.data.panels);
            this.$panels.$groups = this;
        } else {
            this.$panels.active = this.data.active;
        }

        this.data.isConnect = true;
        window.removeEventListener('resize', this._resizeFn);
        window.addEventListener('resize', this._resizeFn);
    }

    disconnectedCallback() {
        // this.removeChild(this.$tabs);
        // this.removeChild(this.$panels);

        this.data.isConnect = false;
        window.removeEventListener('resize', this._resizeFn);
    }

    /**
     * 添加 panel
     * @param {string} name 
     */
    addPanel(name, index) {
        if (index === undefined || index < 0 || index > this.data.panels.length) {
            index = this.data.panels.length;
        }

        if (this.data.panels.indexOf(name) !== -1) {
            return;
        }

        this.data.panels.splice(index, 0, name);

        if (this.$tabs) {
            this.$tabs.addTab(name, index);
        }
        if (this.$panels) {
            this.$panels.addPanel(name, index);
        }
    }

    /**
     * 删除 panel
     * @param {string} name 
     * @param {boolean} isDrop
     */
    async removePanel(name, isDrop) {
        const index = this.data.panels.indexOf(name);

        if (index === -1) {
            return false;
        }

        if (this.$panels) {
            let allow = await this.$panels.removePanel(name, isDrop);
            // 删除 panel 的时候会进行判断是否可以删除
            if (allow === false) {
                return false;
            }
            this.$tabs.removeTab(name);
        }

        this.data.panels.splice(index, 1);

        if (this.data.isConnect) {
            utils.element.emit(this);
        }

        if (this.active === name) {
            this.active = this.data.panels[0] || '';
        }

        // 当 panel 被删除的时候，如果还有 panel 是 loading 状态，一起强制删除
        // 因为长时间处于 loading 状态的 panel 一般来说是不存在的插件面板
        const panels = this.$panels.removeLoadingPanel();
        panels.forEach((name) => {
            const index = this.data.panels.indexOf(name);
            if (index !== -1) {
                this.data.panels.splice(index, 1);
            }

            if (this.active === name) {
                this.active = this.data.panels[0] || '';
            }
        });

        if (this.active === '') {
            setTimeout(() => {
                utils.element.refresh(this);
            }, 200);
        }
    }

    _onFocus(event) {
        event.preventDefault();
        event.stopPropagation();
        this.setAttribute('focused', '');
    }

    _onBlur(event) {
        event.preventDefault();
        event.stopPropagation();
        if (this._focused) {
            return;
        }
        this.removeAttribute('focused');
    }
};

module.exports = DockGroups;
