// @editor/dock/lib/renderer/element/dock-resize.js
'use strict';

const utils = require('../../utils');

let $mask = document.createElement('div');
$mask.style.position = 'absolute';
$mask.style.top = 0;
$mask.style.bottom = 0;
$mask.style.left = 0;
$mask.style.right = 0;
$mask.style.width = '100%';
$mask.style.height = '100%';
$mask.style.zIndex = 999;

// develop style
// $mask.style.background = 'red';
// $mask.style.opacity = 0.1;

/**
 * 切换 mask 的隐藏显示状态
 */
let switchMask = function (bool, direction) {
    $mask.style.cursor = direction === 'column' ? 'ns-resize' : 'ew-resize';
    if (bool) {
        !$mask.parentElement && document.body.appendChild($mask);
    } else {
        $mask.parentElement && document.body.removeChild($mask);
    }
};

const $dragMask = document.createElement('div');
$dragMask.style.background = 'var(--color-focus-fill)';
$dragMask.style.opacity = 0.4;
$dragMask.style.position = 'absolute';
$dragMask.style.zIndex = 999;

class DockResize extends window.HTMLElement {

    constructor () {
        super();

        const cache = {
            direction: '',
            x: 0,
            y: 0,

            top: 0,
            left: 0,
            width: 0,
            height: 0,
        };

        let start = (event) => {
            cache.direction = this.parentElement.getAttribute('direction');
            document.addEventListener('mousemove', move);
            document.addEventListener('mouseup', end);
            switchMask(true, this.parentElement.direction);

            cache.x = event.pageX;
            cache.y = event.pageY;

            const bound = this.getBoundingClientRect();
            cache.top = bound.top;
            cache.left = bound.left;
            cache.width = bound.width;
            cache.height = bound.height;

            document.body.appendChild($dragMask);

            $dragMask.style.top = `${cache.top}px`;
            $dragMask.style.left = `${cache.left}px`;
            if (cache.direction === 'row') {
                $dragMask.style.width = `${0}px`;
                $dragMask.style.height = `${cache.height}px`;
            } else {
                $dragMask.style.width = `${cache.width}px`;
                $dragMask.style.height = `${0}px`;
            }

            // 将可被移动的几个元素快照起来
            utils.layout.snapshot(this, event.pageX, event.pageY);

            this.setAttribute('moving', '');
        };

        let move = (event) => {
            // 移动之前的快照
            // utils.layout.dragLayout({
            //     x: event.pageX, y: event.pageY,
            // });
            const bound = this.getBoundingClientRect();
            if (cache.direction === 'row') {
                const offset = event.pageX - cache.x;
                if (offset > 0) {
                    $dragMask.style.left = `${cache.left}px`;
                    $dragMask.style.width = `${offset}px`;
                    this.style.transform = `translate(${offset}px, 0px)`;
                } else {
                    $dragMask.style.left = `${cache.left + offset}px`;
                    $dragMask.style.width = `${-offset}px`;
                    this.style.transform = `translate(${offset}px, 0px)`;
                }
            } else {
                const offset = event.pageY - cache.y;
                if (offset > 0) {
                    $dragMask.style.top = `${cache.top}px`;
                    $dragMask.style.height = `${offset}px`;
                    this.style.transform = `translate(0px, ${offset}px)`;
                } else {
                    $dragMask.style.top = `${cache.top + offset}px`;
                    $dragMask.style.height = `${-offset}px`;
                    this.style.transform = `translate(0px, ${offset}px)`;
                }
            }
        };

        let end = (event) => {
            // 移动之前的快照
            utils.layout.dragLayout({
                x: event.pageX, y: event.pageY,
            });
            this.style.transform = '';
            $dragMask.parentElement && document.body.removeChild($dragMask);
            document.removeEventListener('mousemove', move);
            document.removeEventListener('mouseup', end);
            switchMask(false, this.parentElement.direction);
            utils.layout.clear();

            this.removeAttribute('moving', '');
        };

        this.addEventListener('mousedown', start);
    }
}

DockResize.line = 2;

module.exports = DockResize;
