// @editor/dock/lib/renderer/element/dock-layout.js
'use strict';

const utils = require('../../utils');

class Data {

    constructor(elem) {
        this._elem = elem;

        this._percent = 1;
        this._minWidth = 300;
        this._minHeight = 200;
        this._minimize = false;
    }

    get minimize() {
        return this._minimize;
    }

    set minimize(val) {
        this._minimize = !!val;

        requestAnimationFrame(() => {
            const event = document.createEvent('HTMLEvents');
            event.initEvent('resize', false, false);
            this._elem.dispatchEvent(event);
        });
    }

    get percent() {
        if (this.minimize) {
            return 0;
        }
        return this._percent;
    }

    set percent(val) {
        this._percent = val;
        this._elem.style.flex = val;
    }

    get minWidth() {
        return this._minWidth;
    }

    set minWidth(val) {
        val = parseInt(val);
        if (isNaN(val)) {
            val = 200;
        }
        this._minWidth = val;
        this._elem.style.minWidth = val + 'px';
    }

    get minHeight() {
        return this._minHeight;
    }

    set minHeight(val) {
        val = parseInt(val);
        if (isNaN(val)) {
            val = 150;
        }
        this._minHeight = val;
        this._elem.style.minHeight = val + 'px';
    }
};

class DockLayout extends window.HTMLElement {

    constructor() {
        super();

        this.data = new Data(this);

        this.addEventListener('panel-area', (event) => {
            if (utils.drag.area(event, this)) {
                event.preventDefault();
                event.stopPropagation();
            }
        });

        this.addEventListener('panel-drop', (event) => {
            if (utils.drag.drop(event, this)) {
                event.preventDefault();
                event.stopPropagation();
            }
        });

        this.addEventListener('panel-leave', (event) => {
            utils.drag.switchMask(false);
        });

        this.addEventListener('resize', () => {
            Array.prototype.forEach.call(this.children, (child) => {
                let event = document.createEvent('HTMLEvents');
                event.initEvent('resize', false, false);
                child.dispatchEvent(event);
            });
        });

        this.addEventListener('move', () => {
            Array.prototype.forEach.call(this.children, (child) => {
                let event = document.createEvent('HTMLEvents');
                event.initEvent('move', false, false);
                child.dispatchEvent(event);
            });
        });
    }

    get direction() {
        return this.getAttribute('direction');
    }

    set direction(val) {
        this.setAttribute('direction', val);
    }

    get minimize() {
        return this.getAttribute('minimize') !== null;
    }

    set minimize(val) {
        if (val) {
            this.setAttribute('minimize', '');
            this.data.minimize = true;
            utils.element.refresh(this);
        } else {
            this.data.minimize = false;
            this.removeAttribute('minimize');
        }
        if (this.children[0] && this.children[0].tagName === 'DOCK-GROUPS') {
            this.children[0].$panels.emit('resize');
        }
    }

    /**
     * 应用一个布局 json
     * @param {object} layout 布局 json
     */
    layout(layout) {
        const panels = this.querySelectorAll('panel-frame');
        if (panels) {
            for (let i = 0; i < panels.length; i++) {
                panels[i].emit('close');
            }
        }
        this.innerHTML = '';

        this.direction = layout.direction || 'none';

        this.data.percent = layout.percent;
        this.data.minWidth = layout['min-width'];
        this.data.minHeight = layout['min-height'];
        if (layout.minimize) {
            this.setAttribute('minimize', '');
        } else {
            this.removeAttribute('minimize');
        }
        this.data.minimize = layout.minimize;

        if (layout.panels) {
            let $groups = document.createElement('dock-groups');
            $groups.panels = layout.panels;
            $groups.active = layout.active || layout.panels[0];
            this.appendChild($groups);
            return;
        }

        if (!layout.children || layout.children.length === 0) {
            return;
        }

        for (let i = 0; i < layout.children.length; i++) {
            if (i !== 0) {
                let $resize = document.createElement('dock-resize');
                this.appendChild($resize);
            }

            let $layout = document.createElement('dock-layout');
            this.appendChild($layout);

            $layout.layout(layout.children[i]);
        }
    }

    dump() {

        let json = {
            'min-width': this.data.minWidth,
            'min-height': this.data.minHeight,
            direction: this.direction,
            percent: this.data._percent,
            minimize: this.data.minimize,
        };

        if (this.children.length === 1) {
            let $child = this.children[0];
            json.active = $child.active;
            json.panels = $child.panels;
        } else {
            let children = [];
            for (let i = 0; i < this.children.length; i++) {
                let child = this.children[i];
                if (child.tagName !== 'DOCK-LAYOUT') {
                    continue;
                }
                children.push(child.dump());
            }
            json.children = children;
        }

        return json;
    }
}

module.exports = DockLayout;
