// @editor/dock/lib/renderer/element/dock-panels.js
'use strict';

const panel = require('@editor/panel');
const utils = require('../../utils');
const { refresh } = require('../../utils/element');

class DockPanels extends window.HTMLElement {

    get panels() {
        return this._panels;
    }

    set panels(val) {
        this._panels = val;
        for (let i = 0; i < val.length; i++) {
            this.addPanel(val[i]);
        }
    }

    get active() {
        return this._active;
    }

    set active(val) {
        const refreshFlags = this._active !== val;

        this._active = val;
        refresh(this._active);
        let $panel = null;
        // 首先隐藏所有面板，然后再显示
        for (let i = 0; i < this.children.length; i++) {
            const $elem = this.children[i];
            if ($elem.name === val) {
                $panel = $elem;
                continue;
            }
            $elem.hidden = true;
        }
        if ($panel) {
            $panel.hidden = false;
            this._$active = $panel;
            // 如果焦点不在面板才需要重新获取焦点
            if ($panel.getAttribute('focused') === null) {
                $panel.focus();
                if (refreshFlags) {
                    this.updateOptions();
                }
            }
        }
    }

    emit(str) {
        const evt = document.createEvent('HTMLEvents');
        evt.initEvent('resize', true, false);
        this._$active.dispatchEvent(evt);
    }

    constructor() {
        super();

        this.addEventListener('dragover', (event) => {
            if (!utils.drag.getDrag()) {
                return;
            }

            event.preventDefault();
            event.stopPropagation();

            let evt = document.createEvent('HTMLEvents');
            evt.pageX = event.pageX;
            evt.pageY = event.pageY;
            evt.initEvent('panel-area', true, false);
            this.dispatchEvent(evt);
        }, true);

        this.addEventListener('drop', (event) => {
            if (!utils.drag.getDrag()) {
                return;
            }

            event.preventDefault();
            event.stopPropagation();

            let evt = document.createEvent('HTMLEvents');
            evt.pageX = event.pageX;
            evt.pageY = event.pageY;
            evt.dataTransfer = event.dataTransfer;
            evt.initEvent('panel-drop', true, false);
            this.dispatchEvent(evt);
        }, true);

        this.addEventListener('dragleave', (event) => {
            if (!utils.drag.getDrag()) {
                return;
            }

            event.preventDefault();
            event.stopPropagation();

            let evt = document.createEvent('HTMLEvents');
            evt.pageX = event.pageX;
            evt.pageY = event.pageY;
            evt.initEvent('panel-leave', true, false);
            this.dispatchEvent(evt);
        }, true);
    }

    addPanel(name, index) {
        const $panels = this.querySelectorAll('panel-frame');
        let $panel = $panels ? Array.prototype.find.call($panels, (item) => {
            return item.name === name;
        }) : null;
        const multiple = $panel && $panel.userData.flags && $panel.userData.flags.multiple;
        if ($panel && !multiple) {
            if (this.active === name) {
                const showEvent = document.createEvent('HTMLEvents');
                showEvent.initEvent('hide', false, false);
                $panel.dispatchEvent(showEvent);

                const hideEvent = document.createEvent('HTMLEvents');
                hideEvent.initEvent('show', false, false);
                $panel.dispatchEvent(hideEvent);
            }
        } else {
            $panel = document.createElement('panel-frame');
            $panel.name = name;
        }

        $panel.hidden = $panel.name !== this.active;

        $panel.addEventListener('load', (event) => {
            this.updateOptions();
            this.$groups.$tabs.addTab(event.target.getAttribute('name'));
        });

        let $children = this.children;
        if (index >= $children.length) {
            this.appendChild($panel);
        } else {
            this.insertBefore($panel, $children[index]);
        }
    }

    /**
     * 关闭一个面板
     * @param {string} name 
     * @param {boolean} isDrop
     */
    async removePanel(name, isDrop) {
        let $panels = this.children;
        for (let i = 0; i < $panels.length; i++) {
            let $panel = $panels[i];

            if ($panel.name === name) {
                let allow = await $panel.close(isDrop);
                if (allow === false) {
                    return false;
                }
                if ($panel.parentElement === this) {
                    this.removeChild($panel);
                }
            }
        }
        this.updateOptions();
        return true;
    }

    /**
     * 强制删除正在加载的面板
     */
    removeLoadingPanel() {
        const panels = [];

        let $panels = this.children;
        for (let i = 0; i < $panels.length; i++) {
            let $panel = $panels[i];

            if (!$panel.isLoaded) {
                i--;
                panels.push($panel.name);

                if ($panel.parentElement === this) {
                    this.removeChild($panel);
                }
            }
        }
        // this.updateOptions();
        return panels;
    }

    updateOptions() {
        let minWidth = 0;
        let minHeight = 0;

        Array.prototype.forEach.call(this.children, ($child) => {
            if ($child.hasAttribute('hidden')) {
                return;
            }
            const size = $child.getData('size');
            minWidth = size ? Math.max(minWidth, size['min-width'] || 0) : 0;
            minHeight = size ? Math.max(minHeight, size['min-height'] || 0) : 0;
        });

        let event = document.createEvent("HTMLEvents");
        event.initEvent('min-size-resize', true, true);
        event.minWidth = minWidth;
        event.minHeight = minHeight;
        this.dispatchEvent(event);
    }
};

module.exports = DockPanels;
