// @cocos/creator-programming-quick-pack/lib/utils/chunk-url.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.quickPackRootURL = exports.FsLocator = exports.isQuickPackURL = void 0;
const asserts_1 = require("@cocos/creator-programming-common/lib/asserts");
const chunk_id_1 = require("./chunk-id");
const relateurl_1 = __importDefault(require("relateurl"));
const url_1 = require("url");
function isQuickPackURL(url) {
    return url.protocol === 'quick-pack:';
}
exports.isQuickPackURL = isQuickPackURL;
function isFsQuickPackURL(url) {
    (0, asserts_1.asserts)(isQuickPackURL(url));
    return url.pathname.startsWith('/file/');
}
class FsLocator {
    constructor(pathname) {
        this._fsRoot = getFsQuickPackURLRoot(pathname.endsWith('/') ? pathname : `${pathname}/`);
        this._relateUrlEvaluator = new relateurl_1.default(this._fsRoot.href);
    }
    getOrigin() {
        return new url_1.URL('.', this._fsRoot);
    }
    toQuickPackURL(chunkId) {
        return (0, chunk_id_1.isRelocatableChunkId)(chunkId)
            ? new url_1.URL(chunkId, this._fsRoot)
            : new url_1.URL(chunkId);
    }
    toChunkId(quickPackURL) {
        (0, asserts_1.asserts)(isQuickPackURL(quickPackURL));
        if (!isFsQuickPackURL(quickPackURL)) {
            return quickPackURL.href;
        }
        else {
            const relUrl = this._relateUrlEvaluator.relate(quickPackURL.href);
            return relUrl;
        }
    }
}
exports.FsLocator = FsLocator;
exports.quickPackRootURL = new url_1.URL('quick-pack:/');
function getFsQuickPackURLRoot(pathname) {
    return new url_1.URL(`/file/${pathname.startsWith('/') ? pathname.substr(1) : pathname}`, exports.quickPackRootURL);
}
