// editor/index.js
'use strict';

exports.App = require('../../lib/dist/app').App;
exports.Dialog = require('../../lib/dist/dialog').Dialog;
exports.EditMode = require('../../lib/dist/edit-mode').EditMode;
exports.I18n = require('../../lib/dist/i18n').I18n;
exports.Message = require('../../lib/dist/message').Message;
exports.Layout = require('../../lib/dist/layout').Layout;
// exports.Logger = require('../../lib/dist/logger').Logger;
exports.Menu = require('../../lib/dist/menu').Menu;
exports.Metrics = require('../../lib/dist/metrics').Metrics;
exports.Network = require('../../lib/dist/network').Network;
exports.Package = require('../../lib/dist/package').Package;
exports.Panel = require('../../lib/dist/panel').Panel;
exports.Profile = require('../../lib/dist/profile').Profile;
exports.Project = require('../../lib/dist/project').Project;
exports.Selection = require('../../lib/dist/selection').Selection;
// exports.Startup = require('../../lib/dist/startup').Startup;
exports.Task = require('../../lib/dist/task').Task;
exports.Theme = require('../../lib/dist/theme').Theme;
exports.UI = require('../../lib/dist/ui-kit').UI;
exports.User = require('../../lib/dist/user').User;
exports.Utils = require('../../lib/dist/utils').Utils;
exports.Module = require('../../lib/dist/module').Module;

const { join } = require('path');

// worker 进程
if (global && global.window && global.window.__name__) {
    process.nextTick(() => {
        // Hack 监听 message 数据
        const mWindows = require(join(exports.App.path, './builtin/messages/static/windows.js'));
        mWindows.load();
    });
}
