// @cocos/ccbuild/modules/build-engine/lib/engine-js/ts-plugins/inline-enum/index.js
"use strict";

var __createBinding = this && this.__createBinding || (Object.create ? function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  var desc = Object.getOwnPropertyDescriptor(m, k);
  if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
    desc = {
      enumerable: true,
      get: function () {
        return m[k];
      }
    };
  }
  Object.defineProperty(o, k2, desc);
} : function (o, m, k, k2) {
  if (k2 === undefined) k2 = k;
  o[k2] = m[k];
});
var __setModuleDefault = this && this.__setModuleDefault || (Object.create ? function (o, v) {
  Object.defineProperty(o, "default", {
    enumerable: true,
    value: v
  });
} : function (o, v) {
  o["default"] = v;
});
var __importStar = this && this.__importStar || function (mod) {
  if (mod && mod.__esModule) return mod;
  var result = {};
  if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
  __setModuleDefault(result, mod);
  return result;
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineEnumTransformer = void 0;
const ts = __importStar(require("@cocos/typescript"));
class EnumInliner {
  constructor(context, enumData) {
    this._context = context;
    this._enumData = enumData;
  }
  visitSourceFile(node, program, context) {
    const result = this.visitNodeAndChildren(node, program, context);
    return result;
  }
  visitNodeAndChildren(node, program, context) {
    return ts.visitEachChild(this.visitNode(node, program), childNode => this.visitNodeAndChildren(childNode, program, context), context);
  }
  visitNode(node, program) {
    if (isAccessExpression(node)) {
      return this.tryCreateLiteral(node, program);
    } else if (ts.isEnumDeclaration(node)) {
      return this.tryReplaceEnumDeclaration(node, program);
    }
    return node;
  }
  tryReplaceEnumDeclaration(node, program) {
    const enumName = node.name;
    const typeChecker = program.getTypeChecker();
    const symbol = typeChecker.getSymbolAtLocation(enumName);
    if (symbol && symbol.declarations && symbol.declarations.length > 0) {
      const declaration = symbol.declarations[0];
      if (ts.isEnumDeclaration(declaration)) {
        let foundAll = true;
        const members = declaration.members.map(member => {
          const memberName = member.name;
          let name = null;
          if (ts.isIdentifier(memberName) || ts.isStringLiteral(memberName)) {
            name = memberName.text;
          } else {
            name = memberName.getText();
          }
          const fullName = `${enumName.text}.${name}`;
          const value = this._enumData.defines[fullName];
          if (value === undefined) {
            foundAll = false;
          }
          return ts.factory.createPropertyAssignment(memberName, typeof value === 'number' ? ts.factory.createNumericLiteral(value) : ts.factory.createStringLiteral(value));
        });
        if (!foundAll) {
          return node;
        }
        const enumObject = ts.factory.createObjectLiteralExpression(members, true);
        const exportModifiers = getModifier(node, ts.SyntaxKind.ExportKeyword) ? [ts.factory.createToken(ts.SyntaxKind.ExportKeyword)] : undefined;
        return ts.factory.createVariableStatement(exportModifiers, ts.factory.createVariableDeclarationList([ts.factory.createVariableDeclaration(enumName, undefined, undefined, enumObject)], ts.NodeFlags.Const));
      }
    }
    return node;
  }
  createNumberOrStringLiteral(value) {
    if (typeof value === 'number') {
      return this._context.factory.createNumericLiteral(value);
    } else if (typeof value === 'string') {
      return this._context.factory.createStringLiteral(value);
    }
    return null;
  }
  findEnumData(name) {
    return this._enumData.defines[name];
  }
  tryCreateLiteral(node, program) {
    const typeChecker = program.getTypeChecker();
    let accessName;
    if (ts.isPropertyAccessExpression(node)) {
      accessName = node.name;
    } else if (ts.isElementAccessExpression(node)) {
      accessName = node.argumentExpression;
    } else {
      return node;
    }
    const symbol = typeChecker.getSymbolAtLocation(accessName);
    // Doesn't support template string now.
    if (node.parent && ts.isTemplateSpan(node.parent)) {
      return node;
    }
    if (symbol && symbol.valueDeclaration) {
      if (symbol.valueDeclaration.kind === ts.SyntaxKind.EnumMember) {
        let expressionName;
        if (ts.isIdentifier(node.expression)) {
          expressionName = node.expression.text;
        } else if (ts.isPropertyAccessExpression(node.expression)) {
          expressionName = node.expression.name.text;
        } else if (ts.isNonNullExpression(node.expression)) {
          if (symbol.parent) {
            expressionName = symbol.parent.name;
          } else {
            console.warn(`NonNullExpression has no parent symbol: ${node.expression.getText()}`);
            return node;
          }
        } else {
          console.warn(`Unsupported expression type: ${node.expression.getText()}, kind: ${node.expression.kind}`);
          return node;
        }
        const enumText = expressionName + '.' + accessName.text;
        const enumInlinedValue = this.findEnumData(enumText);
        if (enumInlinedValue === undefined) {
          return node;
        }
        const inlinedIiteral = this.createNumberOrStringLiteral(enumInlinedValue);
        if (inlinedIiteral) {
          return inlinedIiteral;
        }
      } else if (symbol.valueDeclaration.kind === ts.SyntaxKind.PropertySignature) {
        const accessNameText = accessName.text;
        if (accessNameText === 'BYTES_PER_ELEMENT') {
          const expressionText = node.expression.getText();
          const enumText = expressionText + '.' + accessNameText;
          const enumInlinedValue = this.findEnumData(enumText);
          if (enumInlinedValue === undefined) {
            return node;
          }
          const inlinedIiteral = this.createNumberOrStringLiteral(enumInlinedValue);
          if (inlinedIiteral) {
            return inlinedIiteral;
          }
        }
      }
    }
    return node;
  }
}
function isAccessExpression(node) {
  return ts.isPropertyAccessExpression(node) || ts.isElementAccessExpression(node);
}
function getModifier(node, modifier) {
  return getModifiers(node).find(mod => mod.kind === modifier);
}
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
function getModifiers(node) {
  if (isBreakingTypeScriptApi(ts)) {
    if (!ts.canHaveModifiers(node)) {
      return [];
    }
    return ts.getModifiers(node) || [];
  }
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return node.modifiers || [];
}
function isBreakingTypeScriptApi(compiler) {
  return 'canHaveDecorators' in ts;
}
function inlineEnumTransformer(program, enumData) {
  return context => {
    const inliner = new EnumInliner(context, enumData);
    return file => {
      return inliner.visitSourceFile(file, program, context);
    };
  };
}
exports.inlineEnumTransformer = inlineEnumTransformer;
