// @editor/lib-programming/dist/mod-assembling/quick-pack/utils/chunk-io/writer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChunkWriter = void 0;
const base_1 = require("./base");
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const chunk_id_1 = require("../chunk-id");
const asserts_1 = require("../../../../utils/asserts");
const url_1 = require("url");
const relateurl_1 = __importDefault(require("relateurl"));
class ChunkWriter extends base_1.ChunkIOBase {
    constructor({ origin, chunkRecordFile, sourceCacheDir, importMapFile, logger, }) {
        super({ chunkRecordFile, sourceCacheDir });
        this._origin = origin;
        this._logger = logger;
        this._importMapFile = importMapFile;
    }
    async loadCache() {
        await this._loadChunkRecord();
    }
    async clear() {
        const importMapFile = this._importMapFile;
        try {
            await fs_extra_1.default.unlink(importMapFile);
        }
        catch (err) {
            this._logger.debug(`Failed to delete module record file ${importMapFile}: ${err}`);
        }
        await super.clear();
    }
    getChunkId(url) {
        return chunk_id_1.calculateChunkId(url, this._origin);
    }
    async getChunk(id) {
        return this._chunks[id];
    }
    async addChunk(id, code, map) {
        const physicalLocation = chunk_id_1.calculatePhysicalLocation(id);
        const mTimestamp = Date.now();
        this._chunks[id] = {
            imports: {},
            physicalLocation,
            mTimestamp,
        };
        const sourceCacheFile = this._getChunkSourceFile(physicalLocation);
        // const metadataCacheFile = this._getMetadataCacheFile(qpLocation);
        let sourceMappingURL;
        let serializedMap;
        if (map) {
            serializedMap = typeof map === 'string'
                ? map
                : JSON.stringify(map);
            sourceMappingURL = `${path_1.default.basename(sourceCacheFile)}.map`;
        }
        const codeWithMap = sourceMappingURL
            ? `${code}\n//# sourceMappingURL=${sourceMappingURL}`
            : code;
        await Promise.all([
            // Write metadata
            // fs.outputFile(metadataCacheFile, JSON.stringify(metadata, undefined, 2), { encoding: 'utf8' }),
            // Write code
            fs_extra_1.default.outputFile(sourceCacheFile, codeWithMap, { encoding: 'utf8' }),
            // Write map
            serializedMap
                ? fs_extra_1.default.outputFile(`${sourceCacheFile}.map`, serializedMap, { encoding: 'utf8' })
                : undefined,
        ]);
    }
    async removeChunk(id) {
        delete this._chunks[id];
    }
    async syncChunk(chunkAlias) {
        await this._writeChunkRecord();
        await this._writeImportMap(this._importMapFile, chunkAlias);
    }
    async _writeImportMap(outFile, chunkAlias) {
        const importMapURL = url_1.pathToFileURL(outFile);
        const importMapRelate = new relateurl_1.default(importMapURL.href);
        const getChunkRelativePathFromImportMap = (chunkId) => {
            // Can happen when reload A -> B
            // A is unchanged, but can not load B(since `forceAll: true`)
            // A's dep records are not clear
            asserts_1.asserts(chunkId in this._chunks, 'Something went wrong: module A depends on B, but B is not correctly generated.');
            const chunkSourceFile = this._getChunkSourceFile(this._chunks[chunkId].physicalLocation);
            const relativePath = importMapRelate.relate(url_1.pathToFileURL(chunkSourceFile).href);
            return `./${relativePath}`;
        };
        const importMap = {};
        importMap.imports = {};
        const imports = importMap.imports;
        for (const [alias, chunkId] of Object.entries(chunkAlias)) {
            imports[alias] = getChunkRelativePathFromImportMap(chunkId);
        }
        importMap.scopes = {};
        const scopes = importMap.scopes;
        for (const [chunkId, chunk] of Object.entries(this._chunks)) {
            const importsEntries = Object.entries(chunk.imports);
            if (importsEntries.length === 0) {
                continue;
            }
            const chunkPath = getChunkRelativePathFromImportMap(chunkId);
            const specifierMap = scopes[chunkPath] = {};
            for (const [specifier, resolvedChunkId] of importsEntries) {
                if (typeof resolvedChunkId === 'string') {
                    specifierMap[specifier] = getChunkRelativePathFromImportMap(resolvedChunkId);
                }
                else {
                    specifierMap[specifier] = resolvedChunkId.url;
                }
            }
        }
        await fs_extra_1.default.outputJson(outFile, importMap, { spaces: 2, });
    }
}
exports.ChunkWriter = ChunkWriter;
