// @editor/lib-programming/dist/mod-assembling/quick-pack/utils/chunk-io/writer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChunkWriter = void 0;
const base_1 = require("./base");
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const chunk_id_1 = require("../chunk-id");
class ChunkWriter extends base_1.ChunkIOBase {
    constructor({ origin, chunkRecordFile, sourceCacheDir, }) {
        super({ chunkRecordFile, sourceCacheDir });
        this._origin = origin;
    }
    async loadCache() {
        await this._loadChunkRecord();
    }
    getChunkId(url) {
        return chunk_id_1.calculateChunkId(url, this._origin);
    }
    async getChunk(id) {
        return this._chunks[id];
    }
    async addChunk(id, code, map) {
        const physicalLocation = chunk_id_1.calculatePhysicalLocation(id);
        const mTimestamp = Date.now();
        this._chunks[id] = {
            imports: {},
            physicalLocation,
            mTimestamp,
        };
        const sourceCacheFile = this._getChunkSourceFile(physicalLocation);
        // const metadataCacheFile = this._getMetadataCacheFile(qpLocation);
        let sourceMappingURL;
        let serializedMap;
        if (map) {
            serializedMap = typeof map === 'string'
                ? map
                : JSON.stringify(map);
            sourceMappingURL = `${path_1.default.basename(sourceCacheFile)}.map`;
        }
        const codeWithMap = sourceMappingURL
            ? `${code}\n//# sourceMappingURL=${sourceMappingURL}`
            : code;
        await Promise.all([
            // Write metadata
            // fs.outputFile(metadataCacheFile, JSON.stringify(metadata, undefined, 2), { encoding: 'utf8' }),
            // Write code
            fs_extra_1.default.outputFile(sourceCacheFile, codeWithMap, { encoding: 'utf8' }),
            // Write map
            serializedMap
                ? fs_extra_1.default.outputFile(`${sourceCacheFile}.map`, serializedMap, { encoding: 'utf8' })
                : undefined,
        ]);
    }
    async removeChunk(id) {
        delete this._chunks[id];
    }
    async syncChunk() {
        await this._writeChunkRecord();
    }
}
exports.ChunkWriter = ChunkWriter;
