// @editor/sentry/dist/main/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sentry = void 0;
const Sentry = require("@sentry/electron/main");
const core = require("@sentry/core");
const v_stacks_1 = require("v-stacks");
const ipc = require("@base/electron-base-ipc");
/**
 * 用于上报编辑器报错
 */
class SentryManager {
    _hasInit = false;
    _disabled = false;
    _initInfo;
    _userInfo = null;
    _scope;
    get info() {
        return {
            initInfo: this._initInfo,
            userInfo: this._userInfo,
        };
    }
    get disabled() {
        return this._disabled || !this._hasInit;
    }
    init(info) {
        if (this._hasInit) {
            return;
        }
        this._initInfo = info;
        // 过滤最后一行 sentry 的 log
        function ignoreLoggerStack(stack) {
            const str = stack.split('\n')[1];
            if (str && str.includes('node_modules/@sentry')) {
                return (0, v_stacks_1.ignoreStack)(stack, 1);
            }
            return stack;
        }
        const error = console.error.bind(console);
        console.error = (err, ...args) => {
            err = typeof err === 'string' ? new Error(err) : err;
            const skip = err?.message && info.skipTags.some((tag) => err.message.startsWith(tag));
            if (!skip) {
                Sentry.captureException(err);
            }
            if (err?.stack) {
                // 过滤最后一行 sentry 的 log
                err.stack = ignoreLoggerStack(err.stack);
            }
            error(err, ...args);
        };
        const warn = console.warn.bind(console);
        console.warn = function (item, ...args) {
            if (item && item.stack) {
                // 过滤最后一行 sentry 的 log
                item.stack = ignoreLoggerStack(item.stack);
            }
            warn(item, ...args);
        };
        // https://github.com/cocos-creator/creator-base/commit/976ad5f184a34dc3c207703eb7a02e9dac069f1d
        const normalize = core.normalize;
        // @ts-ignore
        core.normalize = function (input, depth = 100, maxProperties = +Infinity) {
            if (depth === 20 && maxProperties === 2000) {
                depth = 5;
            }
            return normalize(input, depth, maxProperties);
        };
        const { dsn, version, debug, releaseInfo } = info;
        this._scope = Sentry.getCurrentScope();
        Sentry.init({
            dsn: dsn,
            release: version,
            debug: debug,
            sampleRate: 0.5,
            tracesSampleRate: 0.2,
            environment: releaseInfo ? 'production' : 'development',
        });
        if (releaseInfo) {
            this._scope.setExtras({
                'release-date': new Date(releaseInfo.time).toLocaleString('zh-CN', {
                    timeZone: 'Asia/Shanghai',
                }),
                hash: releaseInfo.hash,
            });
        }
        this._hasInit = true;
        this._disabled = false;
        console.debug('init **** success');
    }
    setUserInfo(userInfo = { email: '', nickname: '', cocos_uid: '' }) {
        if (this.disabled) {
            return;
        }
        this._userInfo = userInfo;
        this._scope.setUser({ email: userInfo.email, username: userInfo.nickname, id: userInfo.cocos_uid });
    }
    captureException(...args) {
        if (this.disabled) {
            return;
        }
        return Sentry.captureException(...args);
    }
    close() {
        this._disabled = true;
    }
}
exports.sentry = new SentryManager();
ipc.on('sentry-in-main-is-enable', async (event) => {
    return !exports.sentry.disabled;
});
