// @editor/creator/dist/task/protected/dom.js
"use strict";Object.defineProperty(exports,"__esModule",{value:!0}),exports.show=void 0;const utils_1=require("../utils");class EditorMaskElement extends HTMLElement{$describe;$message;$bar;timeId;time;progress;constructor(){super(),this.attachShadow({mode:"open"}),this.shadowRoot.innerHTML=`
        <style>
            :host > .background {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 0;
                z-index: -1;
                width: 100%;
                height: 100%;
                -webkit-user-drag: none;
            }
            :host > .image {
                display: flex;
                align-items: center;
                height: 83.3vh;
            }
            :host > .image img{
                display: block;
                width: 240px;
                height: 240px;
                opacity: 0.7;
                -webkit-user-drag: none;
            }
            :host > .info {
                width: 33.3vw;
                min-width: 300px;
                height: 16.6vh;
                text-align: center;
                overflow: hidden;
            }
            :host .info > .progress {
                width: 100%;
                height: 4px;
                border-radius: 4px;
                background: #525252ff;
            }
            :host .info > .progress .bar {
                width: 0%;
                height: 100%;
                border-radius: 4px;
                background: #40aacaff;
            }
            :host .info > .describe {
                margin-top: 14px;
                margin-bottom: 7px;
                overflow: hidden;
                white-space: nowrap;
                text-overflow: ellipsis;
            }
            :host .info > .message {
                line-height: 16px;
                color: #525252ff;
                overflow: hidden;
                white-space: nowrap;
                text-overflow: ellipsis;
            }
        </style>
        <img class="background" src="./image/splash-bg.png">
        <div class="image">
            <img src="./image/splash.png">
        </div>
        <div class="info">
            <div class="progress">
                <div class="bar"></div>
            </div>
            <div class="describe">Wait a moment...</div>
            <div class="message"></div>    
        </div>
        `,this.$describe=this.shadowRoot.querySelector(".describe"),this.$message=this.shadowRoot.querySelector(".message"),this.$bar=this.shadowRoot.querySelector(".bar")}show(i,e){this.$describe.innerHTML=i,this.$message.innerHTML=e,this.removeAttribute("hidden"),this.timeId||(this.time=0,this.progress=0,this.timeId=setInterval(()=>{let i=.01;this.time<30?i=2:this.time<60?i=1:this.time<120&&(i=.1),this.progress+=i,this.time++,100<this.progress&&(this.progress=100,clearInterval(this.timeId),this.timeId=void 0),this.$bar.style.width=this.progress+"%"},1e3))}hide(){let i=!1;this.timeId&&(i=!0,clearInterval(this.timeId),this.timeId=void 0),this.progress=100,this.$bar.style.width=this.progress+"%",setTimeout(()=>{this.setAttribute("hidden","")},i?300:0)}}window.customElements.get("editor-mask")||window.customElements.define("editor-mask",EditorMaskElement);const $mask=document.querySelector("#task-sync-mask"),info={title:"",describe:"",message:""};function show(i,e,t){if($mask)if(window.__MAIN__)if(info.title=i||"",info.describe=e=(0,utils_1.translationI18n)(e)||"",info.message=t=(0,utils_1.translationI18n)(t)||"",e=i?e||i:"")$mask.show(e,t);else{if("__InitializationTime__"in window)try{var s=window.__InitializationTime__,o=(window.__InitializationTimeEnd__=Date.now(),window.__InitializationTimeEnd__-s);console.debug("main-window:Initialization: "+o),Editor.Metrics.trackTimeEnd("main-window:Initialization",{value:o}),Editor.Metrics.trackTimeStart("main-window:ready")}catch(i){console.debug(i)}$mask.hide()}else $mask.hide()}exports.show=show;
