// @editor/lib-programming/dist/mod-assembling/quick-pack/utils/chunk-url.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.quickPackRootURL = exports.FsLocator = exports.isQuickPackURL = void 0;
const asserts_1 = require("../../../utils/asserts");
const chunk_id_1 = require("./chunk-id");
const relateurl_1 = __importDefault(require("relateurl"));
function isQuickPackURL(url) {
    return url.protocol === 'quick-pack:';
}
exports.isQuickPackURL = isQuickPackURL;
function isFsQuickPackURL(url) {
    asserts_1.asserts(isQuickPackURL(url));
    return url.pathname.startsWith('/file/');
}
class FsLocator {
    constructor(pathname) {
        this._fsRoot = getFsQuickPackURLRoot(pathname.endsWith('/') ? pathname : `${pathname}/`);
        this._relateUrlEvaluator = new relateurl_1.default(this._fsRoot.href);
    }
    getOrigin() {
        return new URL('.', this._fsRoot);
    }
    toQuickPackURL(chunkId) {
        return chunk_id_1.isRelocatableChunkId(chunkId)
            ? new URL(chunkId, this._fsRoot)
            : new URL(chunkId);
    }
    toChunkId(quickPackURL) {
        asserts_1.asserts(isQuickPackURL(quickPackURL));
        if (!isFsQuickPackURL(quickPackURL)) {
            return quickPackURL.href;
        }
        else {
            const relUrl = this._relateUrlEvaluator.relate(quickPackURL.href);
            return relUrl;
        }
    }
}
exports.FsLocator = FsLocator;
exports.quickPackRootURL = new URL('quick-pack:/');
function getFsQuickPackURLRoot(pathname) {
    return new URL(`/file/${pathname.startsWith('/') ? pathname.substr(1) : pathname}`, exports.quickPackRootURL);
}
