// @editor/lib-programming/dist/mod-assembling/quick-pack/utils/detect-imports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectImports = void 0;
const babel = __importStar(require("@babel/core"));
async function detectImports(ast) {
    const result = [];
    const add = (specifier, loc) => {
        result.push({
            value: specifier,
            loc: loc ? {
                start: { line: loc.start.line, column: loc.start.column },
                end: { line: loc.end.line, column: loc.end.column },
            } : undefined,
        });
    };
    const addStringLiteralSpecifier = (specifier) => {
        add(specifier.value, specifier.loc);
    };
    babel.traverse(ast, {
        ImportDeclaration: (path) => addStringLiteralSpecifier(path.node.source),
        ExportAllDeclaration: (path) => addStringLiteralSpecifier(path.node.source),
        ExportNamedDeclaration: (path) => {
            if (path.node.source) {
                addStringLiteralSpecifier(path.node.source);
            }
        },
        CallExpression: (path) => {
            const { callee, arguments: args } = path.node;
            if (callee.type === 'Import' && args.length === 1) {
                const specifierNode = args[0];
                if (specifierNode.type === 'StringLiteral') {
                    addStringLiteralSpecifier(specifierNode);
                }
                else if (specifierNode.type === 'TemplateLiteral' &&
                    specifierNode.expressions.length === 0 &&
                    specifierNode.quasis.length === 1) {
                    add(specifierNode.quasis[0].value.raw, specifierNode.quasis[0].loc);
                }
            }
        },
    });
    return result;
}
exports.detectImports = detectImports;
