// @editor/lib-programming/dist/executor/pack-mod-instantiation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackModInstantiation = void 0;
const vm_1 = __importDefault(require("vm"));
const asserts_1 = require("../utils/asserts");
const fs_extra_1 = __importDefault(require("fs-extra"));
const i18n_1 = require("../utils/i18n");
const url_1 = require("url");
const defaultPackModuleEvaluator = {
    async evaluate(file) {
        let moduleFileResolved;
        try {
            moduleFileResolved = require.resolve(file);
        }
        catch (err) {
            throw new Error(i18n_1.i18nTranslate('executor_pack_mod_instantiation_error_host_resolve_error', { url: file, reason: err }));
        }
        try {
            const source = await fs_extra_1.default.readFile(moduleFileResolved, 'utf8');
            // This is how Node.js implement require
            const fn = vm_1.default.compileFunction(source, ['System'], {
                filename: url_1.pathToFileURL(moduleFileResolved).href,
            });
            Reflect.apply(fn, undefined, [System]);
        }
        catch (err) {
            throw new Error(i18n_1.i18nTranslate('executor_pack_mod_instantiation_error_host_execute_error', { url: file, reason: err }));
        }
        if (!(moduleFileResolved in require.cache)) {
            console.debug(`${file} resolved to ${moduleFileResolved} is not in module cache!`);
        }
        else {
            const mod = require.cache[moduleFileResolved];
            delete require.cache[moduleFileResolved];
            if (mod && mod.parent) {
                const index = mod.parent.children.indexOf(mod);
                if (index !== -1) {
                    mod.parent.children.splice(index, 1);
                }
            }
        }
    },
};
class PackModInstantiation {
    constructor(quickPackLoader, system, evaluator) {
        this._cachedRegisters = {};
        this._quickPackLoader = quickPackLoader;
        // const system = {
        //     register: (...moduleRegister: ModuleRegister) => {
        //         if (this._lastRegister) {
        //             throw new Error(`Unknown register. Did you have more than one System.register() in your code?`);
        //         }
        //         this._lastRegister = [...moduleRegister];
        //     },
        // };
        // this._registerContext = vm.createContext({
        //     System: system,
        // });
        this._system = system;
        this._evaluator = evaluator !== null && evaluator !== void 0 ? evaluator : defaultPackModuleEvaluator;
    }
    async instantiate(chunkURL) {
        const chunkId = this._quickPackLoader.getChunkId(chunkURL);
        let cached = this._cachedRegisters[chunkId];
        if (!cached) {
            this._cachedRegisters[chunkId] = undefined;
            const mTimestamp = this._quickPackLoader.queryTimestamp(chunkId);
            const resource = await this._quickPackLoader.loadChunkFromId(chunkId);
            asserts_1.asserts(mTimestamp >= 0, `Chunk '${chunkId}' has a bad timestamp.`);
            asserts_1.asserts(resource.type === 'file');
            const file = resource.path;
            // const code = await fs.readFile(file, 'utf8');
            const register = await this._evalCode2(file);
            cached = {
                register,
                mTimestamp,
            };
            this._cachedRegisters[chunkId] = cached;
        }
        return cached.register;
    }
    async lock(maxRetryTime) {
        return this._quickPackLoader.lock(maxRetryTime);
    }
    async unlock() {
        return this._quickPackLoader.unlock();
    }
    /**
     * 重新加载底层的 Pack Loader 并且移除所有失效的缓存。
     */
    async reload() {
        await this._quickPackLoader.reload();
        const cachedChunks = Object.keys(this._cachedRegisters);
        const mTimestamps = this._quickPackLoader.queryTimestamps(cachedChunks);
        cachedChunks.forEach((chunkId, iChunk) => {
            const cache = this._cachedRegisters[chunkId];
            if (!cache || cache.mTimestamp !== mTimestamps[iChunk]) {
                delete this._cachedRegisters[chunkId];
            }
        });
    }
    async getImportMap() {
        return await this._quickPackLoader.loadImportMap();
    }
    async _evalCode2(file) {
        await this._evaluator.evaluate(file);
        const register = this._system.getRegister();
        if (!register) {
            throw new Error(i18n_1.i18nTranslate('executor_system_js_no_module_registered', { url: file }));
        }
        return register;
    }
}
exports.PackModInstantiation = PackModInstantiation;
