// @cocos/creator-programming-quick-pack/lib/proper-lockfile/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkSync = exports.check = exports.unlockSync = exports.unlock = exports.lockSync = exports.lock = void 0;
const lockfile = __importStar(require("./lockfile"));
const adapter_1 = require("./adapter");
async function lock(file, options) {
    const release = await (0, adapter_1.toPromise)(lockfile.lock)(file, options);
    return (0, adapter_1.toPromise)(release);
}
exports.lock = lock;
function lockSync(file, options) {
    const release = (0, adapter_1.toSync)(lockfile.lock)(file, (0, adapter_1.toSyncOptions)(options));
    return (0, adapter_1.toSync)(release);
}
exports.lockSync = lockSync;
async function unlock(file, options) {
    return (0, adapter_1.toPromise)(lockfile.unlock)(file, options);
}
exports.unlock = unlock;
function unlockSync(file, options) {
    return (0, adapter_1.toSync)(lockfile.unlock)(file, (0, adapter_1.toSyncOptions)(options));
}
exports.unlockSync = unlockSync;
function check(file, options) {
    return (0, adapter_1.toPromise)(lockfile.check)(file, options);
}
exports.check = check;
function checkSync(file, options) {
    return (0, adapter_1.toSync)(lockfile.check)(file, (0, adapter_1.toSyncOptions)(options));
}
exports.checkSync = checkSync;
