// @editor/sentry/dist/render/render.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initOptions = exports.sentry = void 0;
const Sentry = require("@sentry/electron/renderer");
const core = require("@sentry/core");
const ipc = require("@base/electron-base-ipc");
/**
 * 用于编辑器渲染进程上报错误
 */
class SentryManager {
    disabled = true;
    init(...args) {
        try {
            const sentryInMainIsEnable = ipc.sendSync('sentry-in-main-is-enable');
            this.disabled = !sentryInMainIsEnable;
            if (this.disabled) {
                return;
            }
            // https://github.com/cocos-creator/creator-base/commit/976ad5f184a34dc3c207703eb7a02e9dac069f1d
            const normalize = core.normalize;
            // @ts-ignore
            core.normalize = function (input, depth = 100, maxProperties = +Infinity) {
                if (depth === 20 && maxProperties === 2000) {
                    depth = 5;
                }
                return normalize(input, depth, maxProperties);
            };
            Sentry.init(...args);
        }
        catch (error) {
            console.error('sentry init error', error);
        }
    }
    captureException(...args) {
        if (this.disabled) {
            return;
        }
        Sentry.captureException(...args);
    }
    captureFeedback(...args) {
        if (this.disabled) {
            return '';
        }
        return Sentry.captureFeedback(...args);
    }
    close() {
        this.disabled = true;
    }
}
exports.sentry = new SentryManager();
exports.initOptions = {
    integrations: [
        Sentry.browserTracingIntegration(),
        // Sentry.replayIntegration(), 会导致内存泄漏
    ],
    // Set tracesSampleRate to 1.0 to capture 100%
    // of transactions for performance monitoring.
    // We recommend adjusting this value in production
    tracesSampleRate: 0.2,
    sampleRate: 0.5
};
