// @editor/build-systemjs/src/index.ts

import * as rollup from 'rollup';
import { terser as rpTerser } from 'rollup-plugin-terser';
import ps from 'path';
// @ts-ignore
import multiEntry from '@rollup/plugin-multi-entry';
import replace from '@rollup/plugin-replace';

const staticDir = ps.join(__dirname, '..', 'static');
const systemJsSourceDir = ps.join(staticDir, 'systemjs-source');

async function build(options: build.Options) {
    const {
        domScriptLoad,
        amd,
    } = options;

    const plugins: rollup.Plugin[] = [
        multiEntry(),
        replace({
            TRACING: 'false',
        }),
    ];
    if (options.terser) {
        plugins.push(rpTerser());
    }

    const inputs = [
        ps.join(systemJsSourceDir, 'src', 'system-core.js'),
    ];
    if (domScriptLoad) {
        inputs.push(ps.join(systemJsSourceDir, 'src', 'features', 'script-load.js'));
    } else {
        inputs.push(ps.join(staticDir, 'cc', 'forbid-script-load.js'));
    }
    inputs.push(
        ps.join(staticDir, 'cc', 'patches.js'),
        ps.join(systemJsSourceDir, 'src', 'extras', 'named-register.js'),
    );
    if (amd) {
        inputs.push(ps.join(systemJsSourceDir, 'src', 'extras', 'amd.js'));
    }

    const rollupBuild = await rollup.rollup({
        input: inputs,
        plugins: plugins,
    });
    await rollupBuild.write({
        format: 'iife',
        strict: false,
        file: options.outFile,
        sourcemap: options.sourceMap,
    });
}

namespace build {
    export interface Options {
        sourceMap: boolean;
        terser: boolean;
        domScriptLoad: boolean;
        amd: boolean;
        outFile: string;
    }
}

export { build };
