// @cocos/ccbuild/modules/build-engine/lib/engine-ts/index.js
"use strict";

var __awaiter = this && this.__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }
  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }
    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }
    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }
    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildTsEngine = void 0;
const engine_builder_1 = require("./engine-builder");
function buildTsEngine(options) {
  var _a;
  return __awaiter(this, void 0, void 0, function* () {
    const builder = new engine_builder_1.EngineBuilder();
    yield builder.build({
      root: options.engine,
      platform: options.platform,
      features: options.features,
      mode: options.mode,
      flagConfig: (_a = options.flags) !== null && _a !== void 0 ? _a : {},
      outDir: options.out
    });
    // NOTE: here we return a fixed result.
    return {
      exports: {},
      chunkAliases: {
        cc: './system-cc.js'
      },
      hasCriticalWarns: false,
      chunkDepGraph: {},
      assetDepGraph: {}
    };
  });
}
exports.buildTsEngine = buildTsEngine;
