// @editor/lib-programming/dist/mod-assembling/mod-lo/mod-lo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mTimestampToString = exports.isEqualMTimestamp = exports.FetchFileError = exports.AccessDeniedError = exports.ModLo = exports.MemoryModule = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const babel = __importStar(require("@babel/core"));
const wrapper_1 = require("./cjs/wrapper");
const detect_1 = require("./cjs/detect");
const url_1 = require("url");
const mod_lo_builtin_mods_1 = require("./utils/mod-lo-builtin-mods");
const asserts_1 = require("../../utils/asserts");
const preset_env_1 = __importDefault(require("@babel/preset-env"));
const perf_hooks_1 = require("perf_hooks");
const url_2 = require("../utils/url");
const logger_1 = require("../utils/logger");
const plugin_annotate_ccclass_1 = __importDefault(require("../../babel/plugin-annotate-ccclass"));
const plugin_detect_circular_1 = __importDefault(require("../../babel/plugin-detect-circular"));
const esm_resolve_1 = require("./resolver/esm-resolve");
const resolve_error_1 = require("./resolver/resolve-error");
const parse_import_map_1 = require("./resolver/parse-import-map");
const cjs_resolve_1 = require("./resolver/cjs-resolve");
const specifier_1 = require("../utils/specifier");
const i18n_1 = require("../../utils/i18n");
const plugin_import_helper_1 = __importDefault(require("../../babel/plugin-import-helper"));
// @ts-expect-error
const plugin_syntax_top_level_await_1 = __importDefault(require("@babel/plugin-syntax-top-level-await"));
class MemoryModule {
    constructor(source) {
        this.source = source;
    }
    get source() {
        return this._source;
    }
    set source(value) {
        this._source = value;
        this._mTimestamp = perf_hooks_1.performance.now();
    }
    get mTimestamp() {
        return this._mTimestamp;
    }
}
exports.MemoryModule = MemoryModule;
/**
 * The uniform Module Loader.
 */
class ModLo {
    constructor({ targets, loose, exportsConditions, guessCommonJsExports, _cc, cr, _internalTransform, _compressUUID, _helperModule = '', importMap, loadMappings, externals, logger, }) {
        this._assetPrefixes = [];
        this._uuidMap = new Map();
        this._loadMappings = [];
        this._memoryModules = {};
        this._conditions = [];
        this._targets = targets;
        this._loose = loose !== null && loose !== void 0 ? loose : false;
        this._guessCjsExports = guessCommonJsExports !== null && guessCommonJsExports !== void 0 ? guessCommonJsExports : false;
        this._cc = _cc;
        this._cr = cr;
        this._internalTransform = _internalTransform;
        this._compressUUID = _compressUUID;
        this._conditions = [...esm_resolve_1.defaultConditions, ...(exportsConditions !== null && exportsConditions !== void 0 ? exportsConditions : [])];
        // TODO: verify bare or URLs
        this._externals = externals !== null && externals !== void 0 ? externals : [];
        this._logger = logger !== null && logger !== void 0 ? logger : logger_1.createLogger({});
        if (importMap) {
            this.setImportMap(importMap.json, importMap.url);
        }
        this._loadMappings.push(...normalizeLoadMappings(mod_lo_builtin_mods_1.modLoBuiltinLoadMappings));
        if (loadMappings) {
            this._loadMappings.push(...normalizeLoadMappings(loadMappings));
        }
        this._loadMappings.sort(([a], [b]) => a > b ? 1 : -1);
        if (_helperModule) {
            this._babelPluginImportHelper = plugin_import_helper_1.default({ helperModule: _helperModule });
        }
    }
    async getMTimestamp(url) {
        const loadURL = this._getLoadURL(url);
        return await this._getMtime(loadURL);
    }
    async load(url) {
        const loadURL = this._getLoadURL(url);
        const source = await this._loadSource(loadURL);
        const moduleType = await this._determinateModuleType(url);
        const metadata = { type: moduleType };
        let code;
        let map;
        if (moduleType === 'esm') {
            ({ code, map } = await this._transform(url, source, undefined));
        }
        else if (moduleType === 'cjs') {
            const cjsInfo = await this._detectCjsInfo(source, url);
            ({ code, map } = await this._transform(url, source, undefined, cjsInfo));
            metadata.cjs = {
                exports: cjsInfo.exports,
                reexports: cjsInfo.reexports,
            };
        }
        else {
            throw new Error(`Unknown module system.`);
        }
        return {
            originalSource: source,
            code,
            map,
            modLoMetadata: metadata,
        };
    }
    async resolve(specifier, from, metadata) {
        const resolved = await this._resolveUnchecked(specifier, from, metadata);
        if (!resolved.isExternal &&
            resolved.url.protocol !== 'node:') { // We do not treat node: URL as unresolved
            const { url } = resolved;
            if (!url.pathname.startsWith('/')) {
                throw new resolve_error_1.CanNotBeBaseURLError(url);
            }
        }
        return resolved;
    }
    async parse(code, map) {
        const inputSourceMap = typeof map === 'string' ? JSON.parse(map) : map;
        const parsed = await babel.parseAsync(code, {
            configFile: false,
            sourceType: 'module',
            inputSourceMap,
            parserOpts: {
                plugins: parserPlugins,
            },
        });
        asserts_1.assertsNonNullable(parsed);
        return parsed;
    }
    /**
     * @param externals Can be bare specifiers or URLs.
     */
    setExternals(externals) {
        this._externals.push(...externals);
    }
    addMemoryModule(url, source) {
        const urlObject = normalizeURLArg(url);
        return this._memoryModules[urlObject.href] = new MemoryModule(source);
    }
    setUUID(url, uuid) {
        assertIsURL(url);
        this._uuidMap.set(url, uuid);
    }
    unsetUUID(url) {
        assertIsURL(url);
        this._uuidMap.delete(url);
    }
    setImportMap(importMap, url) {
        this._parsedImportMap = parse_import_map_1.parseImportMap(importMap, url);
    }
    /**
     * 设置属于 asset 的模块 URL 的前缀。
     * @param prefixes
     */
    setAssetPrefixes(prefixes) {
        this._assetPrefixes = prefixes.slice();
    }
    /**
     * Throws if babel transform failed.
     */
    _throwBabelReturnNull() {
        throw new Error(`Failed to transform.`);
    }
    async _getMtime(url) {
        if (url.href in this._memoryModules) {
            return this._memoryModules[url.href].mTimestamp;
        }
        const filePath = tryConvertAsPath(url);
        if (filePath) {
            let sourceMtime;
            try {
                sourceMtime = (await fs_extra_1.default.stat(filePath)).mtimeMs;
            }
            catch (err) {
                this._logger.debug(`Failed to stat file '${url}' when get mtime:` + err);
                return 0;
            }
            const uuid = this._uuidMap.get(url.href);
            if (uuid) {
                return {
                    mtime: sourceMtime,
                    uuid,
                };
            }
            else {
                return sourceMtime;
            }
        }
        else {
            // TODO: may be we should relax?
            this._logger.debug(`Failed to get mtime of ${url}: not file URL`);
            return 0;
        }
    }
    _getLoadURL(url) {
        if (url.pathname.endsWith('/')) {
            asserts_1.asserts(false, `Module URL shall not ends with /.`);
            return url;
        }
        const href = url.href;
        for (const [prefix, mapped] of this._loadMappings) {
            if (href === prefix) {
                return mapped;
            }
            else if (prefix.endsWith('/') && href.startsWith(prefix)) {
                asserts_1.asserts(mapped.pathname.endsWith('/'));
                return new url_1.URL(href.substr(prefix.length), mapped);
            }
        }
        return url;
    }
    async _loadSource(url) {
        if (url.href in this._memoryModules) {
            return this._memoryModules[url.href].source;
        }
        else if (isCjsInteropUrl(url)) {
            return getCjsInteropSource(url);
        }
        else if (isFileURL(url)) {
            const path = url_1.fileURLToPath(url);
            try {
                return await fs_extra_1.default.readFile(path, 'utf8');
            }
            catch (err) {
                throw new FetchFileError(path, err);
            }
        }
        else {
            throw new AccessDeniedError(url.href);
        }
    }
    async _resolveUnchecked(specifier, from, metadata) {
        if (this._isExternal(specifier)) {
            return {
                specifierOrURL: specifier,
                isExternal: true,
            };
        }
        if (metadata.type === 'esm') {
            const resolved = await this._doEsmResolve(specifier, from, metadata);
            if (!resolved.isExternal) {
                const { url } = resolved;
                if (!isCjsInteropUrl(from) && await this._determinateModuleType(url) === 'cjs') {
                    this._logger.debug(`Create cjs interop url for '${url.href}'`);
                    const cjsInteropURL = createCjsInteropUrl(url);
                    return {
                        isExternal: false,
                        url: cjsInteropURL,
                    };
                }
            }
            return resolved;
        }
        return await this._doCjsResolve(specifier, from, metadata);
    }
    async _doEsmResolve(specifier, from, _metadata) {
        var _a;
        const stdResolved = await esm_resolve_1.esmResolve(specifier, from, {
            importMap: this._parsedImportMap,
            conditions: this._conditions,
        });
        // 对于项目脚本直接的互相导入，在标准解析的基础上，我们允许 extension-less 式的解析。
        const extensionLessResolved = this._isAssetModule(stdResolved) && this._isAssetModule(from)
            ? (_a = await this._tryExtensionLessResolve(stdResolved, ['.ts'])) !== null && _a !== void 0 ? _a : stdResolved : stdResolved;
        if (extensionLessResolved.protocol === 'file:') {
            const path = url_1.fileURLToPath(extensionLessResolved);
            const stat = await this._tryStat(path);
            if (stat.isDirectory()) {
                throw new resolve_error_1.UnsupportedDirectoryImportError();
            }
            else if (!stat.isFile()) {
                throw new resolve_error_1.ModuleNotFoundError(specifier, from);
            }
        }
        // Note: we do not have "post isExternal()"
        return {
            isExternal: false,
            url: extensionLessResolved,
        };
    }
    async _tryExtensionLessResolve(url, extensions) {
        if (await this._fileExists(url)) {
            return url;
        }
        const baseName = url.pathname.split('/').pop();
        if (!baseName || baseName === '') {
            return undefined;
        }
        for (const extension of extensions) {
            let guess;
            if (await this._fileExists(guess = new url_1.URL(`./${baseName}${extension}`, url))) {
                return guess;
            }
        }
        for (const extension of extensions) {
            let guess;
            if (await this._fileExists(guess = new url_1.URL(`./${baseName}/index${extension}`, url))) {
                return guess;
            }
        }
        return undefined;
    }
    async _doCjsResolve(specifier, from, _metadata) {
        // 如果 CommonJS 模块引用了一个不是文件 URL 的 URL，那么直接算解析成功。
        // 因为这里我们要插入 `cjs-loader` 的引用。
        // TODO：或许除了 `cjs-loader` 之外其它的都应该走 legacy resolve？
        if (!specifier_1.isRelativeSpecifier(specifier)) {
            const url = url_2.tryParseURL(specifier);
            if (url && !url_2.hasFileProtocol(url)) {
                return {
                    isExternal: false,
                    url,
                };
            }
        }
        const resolved = await cjs_resolve_1.cjsResolve(specifier, from);
        return {
            isExternal: false,
            url: resolved,
        };
    }
    _isExternal(specifierOrURL) {
        return this._externals.includes(specifierOrURL);
    }
    async _determinateModuleType(file) {
        // https://nodejs.org/api/packages.html#packages_determining_module_system
        switch (true) {
            case !isFileURL(file):
            case file.pathname.endsWith('.mjs'):
            case file.pathname.endsWith('.ts'):
            default:
                return 'esm';
            case file.pathname.endsWith('.cjs'):
                return 'cjs';
            case file.pathname.endsWith('.js'): {
                const packageJson = await esm_resolve_1.readPackageScope(file);
                if ((packageJson === null || packageJson === void 0 ? void 0 : packageJson.type) === 'module') {
                    return 'esm';
                }
                else {
                    return 'cjs';
                }
            }
        }
    }
    async _detectCjsInfo(source, _url) {
        return detect_1.detectCjsInfo(source);
    }
    async _transform(url, source, map, cjs) {
        const inputSourceMap = typeof map === 'string' ? JSON.parse(map) : map;
        const babelPresetEnvOptions = {
            modules: false,
            targets: this._targets,
            loose: this._loose,
        };
        if (this._internalTransform) {
            babelPresetEnvOptions.include = this._internalTransform.includes;
            babelPresetEnvOptions.exclude = this._internalTransform.excludes;
        }
        const babelPresets = [
            [preset_env_1.default, babelPresetEnvOptions],
        ];
        const babelPlugins = [];
        if (cjs) {
            babelPresets.push({
                plugins: [wrapper_1.wrapCjs(cjs.requires, this._guessCjsExports ? cjs.exports : [], this._guessCjsExports ? cjs.reexports : [])],
            });
        }
        else {
            babelPlugins.push(plugin_syntax_top_level_await_1.default);
            babelPresets.push(this._cc);
            const uuid = this._uuidMap.get(url.href);
            if (uuid !== undefined) {
                const baseName = url_2.getBaseName(url);
                const compressedUUID = this._compressUUID(uuid);
                babelPlugins.push([plugin_annotate_ccclass_1.default, {
                        baseName,
                        compressedUUID,
                    }]);
            }
            if (this._cr && this._isAssetModule(url)) {
                babelPlugins.push([plugin_detect_circular_1.default, this._cr]);
            }
        }
        if (!cjs && this._babelPluginImportHelper) {
            babelPlugins.push(this._babelPluginImportHelper);
        }
        const sourceType = cjs ? 'script' : 'module';
        const babelResult = await babel.transformAsync(source, {
            code: false,
            ast: true,
            sourceType,
            inputSourceMap,
            configFile: false,
            presets: babelPresets,
            plugins: babelPlugins,
        });
        asserts_1.asserts(babelResult);
        asserts_1.asserts(babelResult.ast);
        return {
            code: babelResult.ast,
            map: babelResult.map,
        };
    }
    async _tryStat(path) {
        try {
            return await fs_extra_1.default.stat(path);
        }
        catch (_a) {
            return new fs_extra_1.default.Stats();
        }
    }
    async _fileExists(url) {
        return (await this._tryStat(url_1.fileURLToPath(url))).isFile();
    }
    _isAssetModule(url) {
        const href = url.href;
        return this._assetPrefixes.some((assetPrefix) => href.startsWith(assetPrefix));
    }
}
exports.ModLo = ModLo;
class AccessDeniedError extends Error {
    constructor(url) {
        super(i18n_1.i18nTranslate('modLo_access_denied_error', { url }));
    }
}
exports.AccessDeniedError = AccessDeniedError;
class FetchFileError extends Error {
    constructor(file, err) {
        super(i18n_1.i18nTranslate('modLo_fetch_file_error', { path: file, cause: err }));
    }
}
exports.FetchFileError = FetchFileError;
function isEqualMTimestamp(lhs, rhs) {
    if (typeof lhs !== 'object') {
        return lhs === rhs;
    }
    else if (typeof rhs !== 'object') {
        return false;
    }
    else {
        return lhs.mtime === rhs.mtime && lhs.uuid === rhs.uuid;
    }
}
exports.isEqualMTimestamp = isEqualMTimestamp;
function mTimestampToString(mTimestamp) {
    return typeof mTimestamp === 'object'
        ? `${new Date(mTimestamp.mtime)}@${mTimestamp.uuid}`
        : `${new Date(mTimestamp)}`;
}
exports.mTimestampToString = mTimestampToString;
function normalizeURLArg(url) {
    return typeof url === 'string' ? new url_1.URL(url) : url;
}
function assertIsURL(url) {
    asserts_1.asserts(new url_1.URL(url));
}
function isFileURL(url) {
    return url.protocol === 'file:';
}
function tryConvertAsPath(url) {
    if (isCjsInteropUrl(url)) {
        return tryConvertAsPath(getCjsInteropTarget(url));
    }
    else if (isFileURL(url)) {
        return url_1.fileURLToPath(url);
    }
}
function createCjsInteropUrl(url) {
    try {
        const result = new url_1.URL(`?cjs`, url);
        const originalExtension = url_2.replaceExtension(result, '.mjs');
        if (originalExtension) {
            result.searchParams.append('original', originalExtension);
        }
        return result;
    }
    catch (err) {
        throw new resolve_error_1.CjsInteropError(url, err);
    }
}
function isCjsInteropUrl(url) {
    return url.searchParams.has('cjs');
}
function getCjsInteropTarget(url) {
    const target = new url_1.URL(url.href);
    target.searchParams.delete('cjs');
    const originalExtension = target.searchParams.get('original');
    if (originalExtension) {
        target.searchParams.delete('original');
        url_2.replaceExtension(target, originalExtension);
    }
    return target;
}
function getCjsInteropSource(url) {
    const target = getCjsInteropTarget(url);
    return wrapper_1.getCjsInteropModuleSource(`./${target.pathname.split('/').pop()}`);
}
function normalizeLoadMappings(loadMappings) {
    const normalized = [];
    for (const [from, to] of Object.entries(loadMappings)) {
        if (!from.endsWith('/') && to.endsWith('/')) {
            throw new Error(`Bad load mapping: '${from}': '${to}'`);
        }
        const fromURL = new url_1.URL(from);
        const toURL = new url_1.URL(to);
        normalized.push([fromURL.href, toURL]);
    }
    return normalized;
}
const parserPlugins = [
    'asyncGenerators',
    'bigInt',
    'classPrivateMethods',
    'classPrivateProperties',
    'classProperties',
    // 'classStaticBlock',
    'decimal',
    // 'decorators',
    'decorators-legacy',
    // 'doExpressions',
    'dynamicImport',
    // 'estree',
    'exportDefaultFrom',
    // 'exportNamespaceFrom', // deprecated
    // 'flow',
    // 'flowComments',
    // 'functionBind',
    'functionSent',
    'importMeta',
    // 'jsx',
    'logicalAssignment',
    // 'importAssertions',
    'moduleStringNames',
    'nullishCoalescingOperator',
    'numericSeparator',
    'objectRestSpread',
    'optionalCatchBinding',
    'optionalChaining',
    // 'partialApplication',
    // 'pipelineOperator',
    'placeholders',
    // 'privateIn',
    // 'throwExpressions',
    'topLevelAwait',
];
